---
title: "Vignette - detailed explanation of scFeatures"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette_detail}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}

knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


```{r eval=TRUE, include=FALSE, paged.print=TRUE}

library(inline)
library(RhpcBLASctl)
library(S4Vectors)

blas_set_num_threads(1)

```



```{r setup, include = FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```

# Introduction 

This vignette provides a detailed description of scFeatures, including description for each feature category, the expected output and case studies of using the generated features for classification task and survival analysis.   

For a quick starting point (the TLDR version), please see the other vignette "Vignette - summary of scFeatures functions". 

Note that all scFeatures functions contain a `ncores` argument that controls the number of cores. The default is 1 core. 

If you wish to run downstream analysis using the generated features (see "Vignette - association study"), please make sure that when you run scFeatures, the `sample` is in the format of `a_cond_b`. Where `a` is sample ID and `b` is  the condition, eg, `patient09_cond_nonresponder`. This is so that we can retrieve the sample ID and the corresponding condition when running the analysis. 


```{r}
library(scFeatures)
```



# Quick run of scFeatures

scFeatures can be run using one line of code `scfeatures_result <- scFeatures(data)`, which generates a list of dataframes containing all feature types in the form of samples x features.    

```{r eval=FALSE, include=TRUE}
data <- readRDS(system.file("extdata", "example_scrnaseq.rds", package = "scFeatures"))
data <- process_data(data, normalise = TRUE) #  perform normalisation 

scfeatures_result <- scFeatures(data)
```

 

The above function uses all default settings and generate all feature types. We provide an example below that describes how to customise each parameter.   

```{r eval=FALSE, include=TRUE}
data <- readRDS(system.file("extdata", "example_spatial_proteomics.rds", package = "scFeatures"))
data <- process_data(data, normalise = FALSE) # here we don't normalise , because the data has already been normalised 
 
scfeatures_result <- scFeatures(data, 
             # input selective feature types to generate.      
             feature_types = c("proportion_raw", "pathway_gsva", "L_stats", "gene_mean_celltype",    "gene_prop_aggregated") , 
             # by default assumes data is "scrna" (single-cell RNA-seq), this is now set to "spatial_p" (spatial proteomics).   
             type = "spatial_p" ,
             # by default assumes "Homo sapiens", this is now set to `Mus musculus`. 
             species  = "Mus musculus"  ,   
             # by default uses the 50 hallmark genes,  now set to user specified pathways.   
             geneset = list("pathway_a" = c("SMA" ,  "Vimentin"  ),  "pathway_b" = c("B7H3" ,"FoxP3"   )) ,
             # by default uses top variable genes to generate the celltype specific gene expression feature category,now set to user defined genes. 
             celltype_genes = data.frame(celltype = c("Macrophages"  , "Mono/Neu" ) , marker = c("CD3" , "p53")) ,
             # by default uses top variable genes to generate the overall aggregated gene expression feature category, now set to user defined genes.    
             aggregated_genes =  c("CD3" , "p53") , 
             # When passing as SingleCellExperiment or SpatialExperiment, by default we use the assay stored in "logcount" 
             assay = "norm",
            # By default we look for the sample info in "sample" column and the celltype info in "celltype" column
             sample = "imageID", 
             celltype = "cellType",
      #  If users want to construct features from the spatial metrics category, by default we look for the "x_cord" and "y_cord" column 
             spatialCoords = c("x", "y"), 
      # When passing as SingleCellExperiment or SpatialExperiment and the data is spot-based spatial transcriptomics, we also requires a matrix containing cell type prediction probability of each spot 
             spotProbability = predicted_matrix,
             # by default uses single core 
             ncores = 8  )

```



# scFeatures on single-cell RNA-seq data

For demonstration purpose, we provide a subsampled version of the melanoma pre-treatment dataset [1] as used in our manuscript. 

scFeatures takes in data as a Seurat object, with the gene expression stored in the assay `RNA`. The required metadata is the `celltype` and `sample` label. The `condition` label is optional, and is only required if users wish to learn the difference of disease condition using the generated features. 


Also below is an example of how to construct a Seurat object from a count matrix and the corresponding label file.  

```{r}
data_matrix <- readRDS(system.file("extdata", "example_scrnaseq_matrix.rds", 
                                   package = "scFeatures"))
data_meta <- readRDS(system.file("extdata", "example_scrnaseq_label.rds", 
                                 package = "scFeatures"))
data <- Seurat::CreateSeuratObject(data_matrix, meta.data = data_meta)
```

```{r}
data <- process_data(data) # some pre-processing, including normalisation 
 
# confirm the data contain cell types 
unique(data$celltype) 

# confirm the data contain samples  
unique(data$sample)[1:5]

# disease conditions in this data
unique(data$condition)

# format the patient ID as `a_cond_b` eg, `patient09_cond_nonresponder`. This is so that we can retrieve the sample ID and the corresponding condition when running downstream analysis on the generated features 
data$sample <- paste0( data$sample  , "_cond_", data$condition)
unique(data$sample)[1:5]
```


## Feature generation 

Here we demonstrate the usage of scFeatures to extract features from scRNA-seq data. 
Note that, for the `type` argument, the currently supported options are:   
  * `scrna` : which stands for single-cell RNA-seq data 
  * `spatial_p`: which stands for spatial proteomics data  
  * `spatial_t`: which stands for spatial transcriptomics data   

By default, the type is set to `scrna`, therefore it is not necessarily to specify the type when the input data is scRNA-seq data. 


### Cell type proportions 

* Detail:   
For each sample, we calculate the proportion of each cell type:  
  * `Proportion raw`: the untransformed proportion.   
  * `Proportion logit`: the logit-transformed proportion value, which is a common transformation used for proportional data.    
  * `Proportion ratio`: calculates the ratio of proportion between two cell types. 

* Expected output:   
  * `Proportion raw` and `Proportion logit`: the features are in the form of celltype a, celltype b, with the number representing proportions.
  * `Proportion ratio`: the features are in the form of celltype a vs celltype b, celltype a vs celltype c, with the number representing the ratio between the two cell types.  
  
```{r eval=FALSE, include=TRUE}
feature_proportion_raw <- run_proportion_raw(data)
feature_proportion_logit <- run_proportion_logit(data)
feature_proportion_ratio <- run_proportion_ratio(data)
```


### Cell type specific gene expressions


* Detail:   
This feature category aggregate the cells across each cell type and then construct various features. By default, (if the `features` argument is not provided),  we restrict to the top variable genes to reduce the dimensions of the feature. The argument `num_top_gene` determines how many genes to include. 

  * `Gene mean celltype`: the cell type specific gene expression. 
  * `Gene proportion celltype`: for a particular gene, the percentage of cells in a cell type that it is expressed.
  * `Gene correlation celltype`: correlation of expressions of two given genes in a cell type. 

Alternatively, users can provide their own gene of interest to the `features` argument.  

  
* Expected output: 
  * `Gene mean celltype`: the features are in the form of gene 1 celltype a, gene 2 celltype b ... etc,  with the number representing average gene expression of the given  gene across the cells of the the given celltype. 
  * `Gene proportion celltype`: the features are in the form of gene 1 celltype a, gene 2 celltype b ... etc, with the number representing average gene expression of the given gene across the cells of the the given celltype. 
  * `Gene correlation celltype`: the features are in the form of gene 1 vs gene 2 cell type a , gene 1 vs gene 3 cell type b ... etc, with the numbers representing the correlation of the two given genes in the given cell type.   


Note that mitochondria and ribosomal genes may not be as interesting to look at. Therefore, users can choose to remove these genes using the function `remove_mito`. This step is optional.  

```{r eval=FALSE, include=TRUE}
data_remove_mito <- remove_mito(data)

feature_gene_mean_celltype <- run_gene_mean_celltype(data_remove_mito) 

# we can provide our own sets of genes 
genes_of_interest <- data.frame( marker = c("S100A11" , "GZMB" ,  "DUSP1" ) , 
                                celltype = c( "CD8, T Effector Memory" , "CD8, T Effector Memory"  ,  "Naive T Cells") ) 
feature_gene_prop_celltype <- run_gene_prop_celltype(data_remove_mito, features = genes_of_interest)

# we can change the number to pick ~10 top variables genes per cell type
feature_gene_cor_celltype <- run_gene_cor_celltype(data_remove_mito, num_top_gene = 10)
```


```{r eval=TRUE, include=FALSE}
data_remove_mito <- remove_mito(data)

feature_gene_mean_celltype <- run_gene_mean_celltype(data_remove_mito) 
```


### Cell type specific pathway expressions

* Detail:  
This feature category constructs features relating to pathway (gene set) expression. By default (when the `geneset` argument is not specified),  we use the 50 hallmark gene set from msigdb. The users can also provide their geneset of interest in a list format, with each list entry containing a vector of the names of genes in a gene set.   
  * `pathway_gsva`: uses the GSVA function to calculate the gene set enrichment score for individual cells, then aggregates the scores across cells within a cell type. 
  * `pathway_mean`: averages the gene expression of genes in a pathway. 
  * `pathway_prop`: for the genes in a pathway, calculates the percentage of cells that these genes are expressed in each cell type.  
  
* Expected output: 
  * `pathway_gsva` the features are in the form of pathway 1 celltype a, pathway 2 celltype b ... etc, with the number representing the gene set enrichment score of a given pathway in cells from a given celltype.  
  * `pathway_mean` the features are in the form of pathway 1 celltype a, pathway 2 celltype b ... etc, with the number representing the averaged expression of a given pathway in cells from a given celltype.  
  * `pathway_prop` the features are in the form of pathway 1 celltype a, pathway 2 celltype b ... etc, with the number representing the proportion of expression of a given pathway in cells from a given celltype.  



Note that it is necessarily to indicate whether the species is "Homo sapiens" or "Mus musculus". The default is "Homo sapiens".  


```{r eval=FALSE, include=TRUE}
# This function can take a while. Therefore for larger datasets (eg, over 30,000 cells), we provide a subsample function to increase the speed. 
feature_pathway_gsva <- run_pathway_gsva(data, geneset = NULL, species = "Homo sapiens", subsample = FALSE)
feature_pathway_mean <- run_pathway_mean(data)
feature_pathway_prop <- run_pathway_prop(data)
```



### Overall aggregated gene expressions 

* Detail: 
The feature types in this category is calculated based on the aggregated expressions across all cells. The argument `num_top_gene` determines how many genes to include. By default (if the `features` argument is not provided), the algorithm picks the top variable genes.

* Expected output:   
  * `Gene mean`:  The features are in the form of gene 1, gene 2 ... etc, with the numbers representing averaged gene expression across all cells.  
  * `Gene cor`:  The features are in the form of gene 1 vs gene 2, gene 1 vs gene 3 ... etc, with the numbers representing correlation of gene expressions.   
  * `Gene prop`:  The features are in the form of gene 1, gene 2 ... etc, with the numbers representing the proportion that the gene is expressed across all cells.  


```{r eval=FALSE, include=TRUE}
feature_gene_mean_aggregated <- run_gene_mean(data)

genes_of_interest <- c("TIGIT", "PDCD1")
feature_gene_cor_aggregated  <- run_gene_cor(data, features = genes_of_interest)
# we can change this number to pick 100 top variable genes 
feature_gene_prop_aggregated  <- run_gene_prop(data, num_top_gene = 100)
```



## Classification of conditions using the generated features

scFeatures provides a wrapper around the classification function in ClassifyR (https://www.bioconductor.org/packages/release/bioc/html/ClassifyR.html) to build disease prediction model from the generated features. 


The output from scFeatures is a matrix of sample x feature, ie, the row corresponds to each sample, the column corresponds to the feature, and can be directly used as the `X`. The order of the rows is in the order of unique(data$sample). 

Here we use the `condition` as the `y` to build classification model on the disease condition. 

```{r eval=TRUE, include=TRUE}
# Using the feature type gene mean celltype as an example")
 
# inspect the first 5 rows and first 5 columns
feature_gene_mean_celltype[1:5, 1:5]

# inspect the dimension of the matrix
# this means that there are 12354 features in this feature type
dim(feature_gene_mean_celltype)

# confirm that the rows in in the order of unique(data$sample)
all(rownames(feature_gene_mean_celltype) == unique(data$sample))

```


We recommend using `ClassifyR` to do cross-validated classification with the extracted feaures. 

```{r eval=TRUE, include=TRUE, fig.height=4, fig.width=7}
# X is the feature type generated
# y is the condition for classification
X <- feature_gene_mean_celltype
y <- data@meta.data[!duplicated(data$sample), ]
y <- y[match(rownames(X), y$sample), ]$condition

# run the classification model using random forest
result <- ClassifyR::crossValidate(
    X, y,
    classifier = model, nCores = ncores,
    nFolds = 3, nRepeats = 20
  )

ClassifyR::performancePlot(results = result)
```
 

# scFeatures on spatial proteomics data

For demonstration purpose, we provide a subsampled version of triple negative breast cancer patients dataset [2] as used in our manuscript.  

For spatial proteomics, the required information are `sample`, `x_cord` and `y_cord` (the x- and y- coordinates of the cell).  


```{r}
data <- readRDS(system.file("extdata", "example_spatial_proteomics.rds", 
                                   package = "scFeatures"))

# here we don't normalise , because the data has already been normalised 
data <- process_data(data, normalise = FALSE)

# For spatial proteomics, we require the x- and y- coordinates  
data$x_cord[1:5]
```



## Feature generation

### Cell type specific proportions

For spatial proteomics, we need to set the `type` to `spatial_p`

```{r eval=FALSE, include=TRUE}
feature_proportion_raw <- run_proportion_raw(data, type = "spatial_p")
feature_proportion_logit <- run_proportion_logit(data, type = "spatial_p")
feature_proportion_ratio <- run_proportion_ratio(data, type = "spatial_p")
```


### Cell type specific gene expressions 

By default, this feature category calculates the gene/protein expression based on the top 100 most variable genes/proteins per cell type. However, because this data only measured 38 proteins, all proteins will be used.  

```{r eval=FALSE, include=TRUE}
feature_gene_mean_celltype <- run_gene_mean_celltype(data,  type = "spatial_p")
feature_gene_prop_celltype <- run_gene_prop_celltype(data, type = "spatial_p")
feature_gene_cor_celltype <- run_gene_cor_celltype(data, type = "spatial_p")
```


### Overall aggregated gene expressions 

By default, this feature category calculates the gene/protein expression based on the top 1500 most variable genes/proteins. However, because in this data there are only 38 proteins, we will use all proteins.  

```{r eval=FALSE, include=TRUE}
feature_gene_mean_aggregated <- run_gene_mean(data, type = "spatial_p")
feature_gene_prop_aggregated  <- run_gene_prop(data, type = "spatial_p")
feature_gene_cor_aggregated  <- run_gene_cor(data, type = "spatial_p")
```


### Spatial metrics 

* Detail:   
This feature category is designed for spatial data and contains feature types commonly used for spatial analysis. 
  * `L_function`: the L values between the pairs of proteins are calculated using the L function defined in literature. L value greater than zero indicates spatial attraction of the pair of proteins whereas L value less than zero indicates spatial repulsion. 
  * `Morans_I`:  Moran’s I are calculated using the function defined in literature. It calculates the spatial autocorrelation based on both the locations and values simultaneously. A value closer to 1 indicates clustering of similar values and a value closer to -1 indicates clustering of dissimilar values. A value of 0 indicates no particular clustering structure, ie, the values are spatially distributed randomly.   
  * `celltype_interaction`:  We find the nearest neighbours of each cell and the cell types of these neighbours. These are considered as spatial interaction pairs. The cell type composition of the spatial interaction pairs are used as features.  
  * `nn_correlation`:  Pearson's correlation is calculated for the protein expression between a cell with its nearest neighbour cell for spatial proteomics.  

* Expected output: 
  * `L_function`:  The features are in the form of protein 1 vs protein 2, protein 1 vs protein 3 ... etc, with the numbers representing the L values.  
  * `Morans_I`:  The features are in the form of protein 1, protein 2 ... etc, with the numbers representing Moran's value. 
  * `celltype_interaction`: The features are in the form of protein 1 vs protein 2, protein 1 vs protein 3 ... etc, with the numbers representing the proportion of each interaction pairs in a give sample. 
  * `nn_correlation`:  The features are in the form of protein 1, protein 2 ... etc, with the numbers representing Pearson's correlation.

```{r eval=FALSE, include=TRUE}
feature_L_stats <- run_L_function(data, type = "spatial_p")
feature_morans_I <- run_Morans_I(data, type = "spatial_p")
feature_celltype_interaction <- run_celltype_interaction( data, type = "spatial_p")
feature_nn_correlation <- run_nn_correlation( data, type = "spatial_p")
```
 
```{r eval=TRUE, include=FALSE}
feature_nn_correlation <- run_nn_correlation( data, type = "spatial_p")
```



## Survival analysis using the generated features

Given this dataset contain survival outcome of the patients, we can perform survival analysis using the generated features. 

```{r eval=TRUE, fig.height=5, fig.width=7, include=TRUE}
library(survival)
library(survminer)

# get meta data of the patients with stage 2 breast cancer 
meta_patients <- data@meta.data
meta_patients <- meta_patients[!duplicated(meta_patients$sample), ]
meta_patients <- meta_patients[  which(meta_patients$STAGE == "2") , ]


nncor <- t(feature_nn_correlation)
nncor <- nncor[, as.character(meta_patients$sample)]

# run hierarchical clustering 
hclust_res <- hclust(as.dist(1-cor(nncor, method="pearson")), method="ward.D2")
cluster_res <- cutree(hclust_res, k = 2)
meta_patients$cluster <- factor(cluster_res)

# plot survival curve 
fit <- survfit(Surv(Survival_days_capped_2016.1.1, Censoring) ~ cluster, data = meta_patients)
ggsurv <- ggsurvplot(fit, conf.int = FALSE, risk.table = TRUE, 
                     risk.table.col="strata", pval = TRUE)
ggsurv

```



# scFeatures on spatial transcriptomics data

For demonstration purpose, we provide the subsampled version of ALS dataset [3] as used in our manuscript.  
 
For spatial transcriptomics, the required information are `sample`, `x_cord` and `y_cord` (the x- and y- coordinates of each spot). 

Additionally, there are two assays needed, the `RNA`, which contains the gene expression of each spot and the `predictions`. 

The `predictions` assay is a matrix in the form of cell types x spot, which stores the cell type probability of each spot. This can be obtained from performing cell type prediction using reference data, for example, using `SCTransform` from Seurat (see https://satijalab.org/seurat/articles/spatial_vignette.html). The purpose of having this cell type probability matrix is due to spatial transcriptomics assay contain multiple cells from multiple cell type populations, and this cell type prediction is necessary for generating cell type specific feature types. 


```{r eval=FALSE, include=TRUE}

data <- readRDS(system.file("extdata", "example_spatial_transcriptomics.rds", 
                                   package = "scFeatures"))
data <- process_data(data, normalise = FALSE) # data already normalised

# for spatial transcriptomics, we need to estimate the number of cells in each spot
data <- get_num_cell_per_spot(data)

# how the prediction assay look like 
# the rows contain the predicted cell type probability of each spot
data@assays$predictions[, 1] 

# confirm the data include samples  
unique(data$sample)[1:5]

# conditions in this data
unique(data$condition)

# format the patient ID as `a_cond_b` eg, `patient09_cond_nonresponder`. This is so that we can retrieve the sample ID and the corresponding condition when running downstream analysis on the generated features 
data$sample <- paste0( data$sample  , "_cond_", data$condition)
```

 

## Feature generation

For spatial transcriptomics, each spot contains multiple cells and potentially from multiple cell types. Therefore, the implementation for the feature categories involving cell type specific features (i.e., cell type proportions, cell type specific gene expressions, cell type specific pathway expressions, spatial metrics) are different to their implementations for the single-cell based datasets (i.e., scRNA-seq and spatial proteomics) and described in details below.  


### Cell type proportions 

* Detail: While in spatial transriptomics, the measurement at each spot is taken from multiple cells, we can use the cell type probability matrix and the relative number of cells in each spot jointly to estimate the cell type proportion in each sample. 

* Expected output:  
The output format is the same as for scRNA-seq data.   

```{r eval=FALSE, include=TRUE}
feature_proportion_raw <- run_proportion_raw(data, type = "spatial_t")
feature_proportion_logit <- run_proportion_logit(data, type = "spatial_t")
feature_proportion_ratio <- run_proportion_ratio(data, type = "spatial_t")
```


### Cell type specific gene expressions 

* Detail:  We use the cell type probability to obtain the regression coefficients of each gene associated with each cell type. 

* Expected output: For each sample, the features are in the form of gene a celltype a, gene a celltype b ... etc, with the number representing regression coefficient.     


```{r eval=FALSE, include=TRUE }
data_remove_mito <- remove_mito(data)

feature_gene_mean_celltype <- run_gene_mean_celltype(data_remove_mito, type = "spatial_t")
```


### Cell type specific pathway expressions 

* Detail: We use the approach above to obtain the  regression coefficient of each cell type for each gene. The regression coefficients for all the genes involved in a particular pathway are then summed.   

* Expected output: For each sample, the features are in the form of pathway a celltype a, pathway a celltype b ... etc, with the number representing the summation of regression coefficient.  


```{r eval=FALSE, include=TRUE }
feature_pathway_mean <- run_pathway_mean(data, geneset = NULL, species = "Mus musculus", 
                                         type =  "spatial_t")
```


### Overall aggregated gene expressions

* Detail:  
This is based on the expression of genes across all cells.  

* Expected output:   
The output format is the same as for scRNA-seq data.   

```{r  eval=FALSE, include=TRUE }
feature_gene_mean_aggregated  <- run_gene_mean(data, type = "spatial_t")
feature_gene_prop_aggregated  <- run_gene_prop(data, type = "spatial_t")
feature_gene_cor_aggregated  <- run_gene_cor(data, type = "spatial_t")
```


### Spatial metrics 

* Detail:   
This feature category is designed for spatial data and contains feature types commonly used for spatial analysis. 
  * `L_function`: the L values between the pairs of genes are calculated using the L function defined in literature. We used the estimated cell type proportion in each spot to calculate the L function.   
  * `Morans_I`:  The implementation is the same as for spatial proteomics. 
  * `celltype_interaction`:  We assume that the nearest neighbours should be the cells captured within each spot and consider them as the spatial interaction pairs. We then use the estimated cell type proportion in each spot to calculate the spatial interaction between cell types.  
  * `nn_correlation`:  Pearson's correlation is calculated for the gene expression between a spot with its nearest neighbour spot.  

* Expected output: 
  * `L_function`:  The features are in the form of gene 1 vs gene 2, gene 1 vs gene 3 ... etc, with the numbers representing the L values.  
  * `Morans_I`:  The features are in the form of gene 1, gene 2 ... etc, with the numbers representing Moran's value. 
  * `celltype_interaction`: The features are in the form of gene 1 vs gene 2, gene 1 vs gene 3 ... etc, with the numbers representing the proportion of each interaction pairs in a give sample. 
  * `nn_correlation`:  The features are in the form of gene 1, gene 2 ... etc, with the numbers representing Pearson's correlation.

```{r eval=FALSE, include=TRUE}
feature_L_stats <- run_L_function(data, type = "spatial_t")
feature_morans_I <- run_Morans_I(data, type = "spatial_t")
feature_celltype_interaction <- run_celltype_interaction(data, type = "spatial_t")
feature_nn_correlation <- run_nn_correlation(data, type = "spatial_t")
```
 
 
# Reference 

1. Defining T cell states associated with response to checkpoint immunotherapy in melanoma. Cell, 175(4), 998-1013.    
2. Keren et al. (2019). MIBI-TOF: A multiplexed imaging platform relates cellular phenotypes and tissue structure. Science advances, 5(10), eaax5851.  
3. Maniatis et al. (2019). Spatiotemporal dynamics of molecular pathology in amyotrophic lateral sclerosis. Science, 364(6435), 89-93.  


# sessionInfo()

```{r}
sessionInfo()
```