<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Vignette - summary of scFeatures functions</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Vignette - summary of scFeatures
functions</h1>



<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>This vignette provide a list of all functions in scFeatures for quick
reference. For detailed explanation of scFeatures and case studies,
please refer to the other vigentte “Vignette - detailed explanation of
scFeatures”.</p>
<p>If you wish to run downstream analysis using the generated features
(see “Vignette - association study”), please make sure that when you run
scFeatures, the <code>sample</code> is in the format of
<code>a_cond_b</code>. Where <code>a</code> is sample ID and
<code>b</code> is the condition, eg,
<code>patient09_cond_nonresponder</code>. This is so that we can
retrieve the sample ID and the corresponding condition when running the
analysis.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scFeatures)</span></code></pre></div>
</div>
<div id="quick-run-of-scfeatures" class="section level1">
<h1>Quick run of scFeatures</h1>
<p>scFeatures can be run using one line of code
<code>scfeatures_result &lt;- scFeatures(data)</code>, which generates a
list of dataframes containing all feature types in the form of samples x
features.</p>
<p>For input data format, refer to sections below.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_scrnaseq.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;scFeatures&quot;</span>))</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">process_data</span>(data, <span class="at">normalise =</span> T) <span class="co">#  perform normalisation </span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>scfeatures_result <span class="ot">&lt;-</span> <span class="fu">scFeatures</span>(data)</span></code></pre></div>
<p>The above function uses all default settings and generate all feature
types. We provide an example below on customising each of the
parameters.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_spatial_proteomics.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;scFeatures&quot;</span>))</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">process_data</span>(data, <span class="at">normalise =</span> F) <span class="co"># here we don&#39;t normalise , because the data has already been normalised </span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>scfeatures_result <span class="ot">&lt;-</span> <span class="fu">scFeatures</span>(data, </span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>             <span class="co"># input selective feature types to generate.      </span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">feature_types =</span> <span class="fu">c</span>(<span class="st">&quot;proportion_raw&quot;</span>, <span class="st">&quot;pathway_gsva&quot;</span>, <span class="st">&quot;L_stats&quot;</span>, <span class="st">&quot;gene_mean_celltype&quot;</span>,</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>                               <span class="st">&quot;gene_prop_aggregated&quot;</span>) , </span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>             <span class="co"># by default assumes `scrna`, this is now set to `spatial_p`.   </span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">type =</span> <span class="st">&quot;spatial_p&quot;</span> ,</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>             <span class="co"># by default assumes `Homo sapiens`, this is now set to `Mus musculus`.       </span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">species  =</span> <span class="st">&quot;Mus musculus&quot;</span>  ,   </span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>             <span class="co"># by default uses the 50 hallmark genes,  now set to user specified pathways.   </span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>             <span class="at">geneset =</span> <span class="fu">list</span>(<span class="st">&quot;pathway_a&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;SMA&quot;</span> ,  <span class="st">&quot;Vimentin&quot;</span>  ),  <span class="st">&quot;pathway_b&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;B7H3&quot;</span> ,<span class="st">&quot;FoxP3&quot;</span>   )) ,</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>             <span class="co"># by default uses top variable genes to generate the celltype specific gene expression feature category,now set to user defined genes. </span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>             <span class="at">celltype_genes =</span> <span class="fu">data.frame</span>(<span class="at">celltype =</span> <span class="fu">c</span>(<span class="st">&quot;Macrophages&quot;</span>  , <span class="st">&quot;Mono/Neu&quot;</span> ) , <span class="at">marker =</span> <span class="fu">c</span>(<span class="st">&quot;CD3&quot;</span> , <span class="st">&quot;p53&quot;</span>)) ,</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>             <span class="co"># by default uses top variable genes to generate the overall aggregated gene expression feature category, now set to user defined genes.    </span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>             <span class="at">aggregated_genes =</span>  <span class="fu">c</span>(<span class="st">&quot;CD3&quot;</span> , <span class="st">&quot;p53&quot;</span>) , </span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>             <span class="co"># by default uses single core </span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>             <span class="at">ncores =</span> <span class="dv">8</span>  )</span></code></pre></div>
<p>Below we list the function for generating each individual feature
types for reference.</p>
</div>
<div id="scfeatures-for-single-cell-rna-seq-data" class="section level1">
<h1>scFeatures for single-cell RNA-seq data</h1>
<p>The input data is a Seurat object containing <code>celltype</code>
and <code>sample</code> label.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co"># read in data </span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_scrnaseq.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;scFeatures&quot;</span>))</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">process_data</span>(data, <span class="at">normalise =</span> T) <span class="co">#  perform normalisation </span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co"># format the patient ID as `a_cond_b` eg, `patient09_cond_nonresponder`. This is so that we can retrieve the sample ID and the corresponding condition when running downstream analysis on the generated features </span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>sample <span class="ot">&lt;-</span> <span class="fu">paste0</span>( data<span class="sc">$</span>sample  , <span class="st">&quot;_cond_&quot;</span>, data<span class="sc">$</span>condition)</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="co"># cell type proportions</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>feature_proportion_raw <span class="ot">&lt;-</span> <span class="fu">run_proportion_raw</span>(data)</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>feature_proportion_logit <span class="ot">&lt;-</span> <span class="fu">run_proportion_logit</span>(data)</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>feature_proportion_ratio <span class="ot">&lt;-</span> <span class="fu">run_proportion_ratio</span>(data)</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a><span class="co"># cell type specific gene expressions</span></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>data_remove_mito <span class="ot">&lt;-</span> <span class="fu">remove_mito</span>(data) <span class="co"># optional step, run if mitochondria amd ribosomal genes are not of interest </span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a><span class="co"># by default, use the top variable genes</span></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a><span class="co"># users can change the number of top variable genes by changing `num_top_gene` </span></span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a><span class="co"># alternatively, users can also input their genes of interest in a dataframe format</span></span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>genes_of_interest <span class="ot">&lt;-</span> <span class="fu">data.frame</span>( <span class="at">marker =</span> <span class="fu">c</span>(<span class="st">&quot;S100A11&quot;</span> , <span class="st">&quot;GZMB&quot;</span> ,  <span class="st">&quot;DUSP1&quot;</span> ) , </span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>                                <span class="at">celltype =</span> <span class="fu">c</span>( <span class="st">&quot;CD8, T Effector Memory&quot;</span> , <span class="st">&quot;CD8, T Effector Memory&quot;</span>  ,  <span class="st">&quot;Naive T Cells&quot;</span>) ) </span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>feature_gene_mean_celltype <span class="ot">&lt;-</span> <span class="fu">run_gene_mean_celltype</span>(data_remove_mito)  </span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>feature_gene_prop_celltype  <span class="ot">&lt;-</span> <span class="fu">run_gene_prop_celltype</span>(data_remove_mito, <span class="at">genes =</span> genes_of_interest)</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>feature_gene_cor_celltype <span class="ot">&lt;-</span> <span class="fu">run_gene_cor_celltype</span>(data_remove_mito,  <span class="at">num_top_gene =</span> <span class="dv">50</span>)</span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a><span class="co"># by default, pick around 100 most variable genes per cell type, can change this number through the num_top_gene argument </span></span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a><span class="co"># cell type specific pathway expressions </span></span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a>feature_pathway_gsva <span class="ot">&lt;-</span> <span class="fu">run_pathway_gsva</span>(data, <span class="at">species =</span> <span class="st">&quot;Homo sapiens&quot;</span>) </span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a>feature_pathway_mean <span class="ot">&lt;-</span> <span class="fu">run_pathway_mean</span>(data , <span class="at">species =</span> <span class="st">&quot;Homo sapiens&quot;</span>)</span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a><span class="co"># by default, use the 50 hallmark pathways, users can also input their gene set of interest in a list format </span></span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a>geneset <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="st">&quot;pathway_a&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;CAPNS1&quot;</span>, <span class="st">&quot;TLCD1&quot;</span>), </span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a>                <span class="st">&quot;pathway_b&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;PEX6&quot;</span>,<span class="st">&quot;DPRXP4&quot;</span> ))  </span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a>feature_pathway_prop <span class="ot">&lt;-</span> <span class="fu">run_pathway_prop</span>(data, <span class="at">geneset=</span>geneset, <span class="at">species =</span> <span class="st">&quot;Homo sapiens&quot;</span>)</span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a><span class="co"># cell type specific cell-cell interactions</span></span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a>feature_CCI <span class="ot">&lt;-</span> <span class="fu">run_CCI</span>(data , <span class="at">species =</span> <span class="st">&quot;Homo sapiens&quot;</span> )</span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a><span class="co"># Bulk expressions </span></span>
<span id="cb4-54"><a href="#cb4-54" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb4-55"><a href="#cb4-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-56"><a href="#cb4-56" aria-hidden="true" tabindex="-1"></a><span class="co"># by default, use the top variable genes</span></span>
<span id="cb4-57"><a href="#cb4-57" aria-hidden="true" tabindex="-1"></a><span class="co"># users can change the number of top variable genes by changing `num_top_gene` </span></span>
<span id="cb4-58"><a href="#cb4-58" aria-hidden="true" tabindex="-1"></a><span class="co"># alternatively, users can also input their genes of interest in a list format</span></span>
<span id="cb4-59"><a href="#cb4-59" aria-hidden="true" tabindex="-1"></a>feature_gene_mean_bulk <span class="ot">&lt;-</span> <span class="fu">run_gene_mean</span>(data )</span>
<span id="cb4-60"><a href="#cb4-60" aria-hidden="true" tabindex="-1"></a>feature_gene_cor_bulk <span class="ot">&lt;-</span> <span class="fu">run_gene_cor</span>(data ,  <span class="at">num_top_gene =</span> <span class="dv">5</span> )</span>
<span id="cb4-61"><a href="#cb4-61" aria-hidden="true" tabindex="-1"></a>genes_of_interest <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;TIGIT&quot;</span>, <span class="st">&quot;PDCD1&quot;</span>)</span>
<span id="cb4-62"><a href="#cb4-62" aria-hidden="true" tabindex="-1"></a>feature_gene_prop_bulk <span class="ot">&lt;-</span> <span class="fu">run_gene_prop</span>(data, <span class="at">genes =</span> genes_of_interest )</span>
<span id="cb4-63"><a href="#cb4-63" aria-hidden="true" tabindex="-1"></a><span class="co"># by default, pick 1500 most variable genes, can change this number through the num_top_gene argument </span></span>
<span id="cb4-64"><a href="#cb4-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-65"><a href="#cb4-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-66"><a href="#cb4-66" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb4-67"><a href="#cb4-67" aria-hidden="true" tabindex="-1"></a><span class="co"># run classification model on the generated features </span></span>
<span id="cb4-68"><a href="#cb4-68" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb4-69"><a href="#cb4-69" aria-hidden="true" tabindex="-1"></a>X <span class="ot">&lt;-</span> feature_gene_mean_celltype</span>
<span id="cb4-70"><a href="#cb4-70" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> data<span class="sc">@</span>meta.data[<span class="sc">!</span><span class="fu">duplicated</span>(data<span class="sc">$</span>sample),  ]</span>
<span id="cb4-71"><a href="#cb4-71" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> y[ <span class="fu">match</span>(<span class="fu">rownames</span>(X), y<span class="sc">$</span>sample), ]<span class="sc">$</span>condition</span>
<span id="cb4-72"><a href="#cb4-72" aria-hidden="true" tabindex="-1"></a>feature_performance <span class="ot">&lt;-</span> <span class="fu">run_classification</span>( X, y, <span class="at">model =</span> <span class="st">&quot;randomforest&quot;</span>, </span>
<span id="cb4-73"><a href="#cb4-73" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">ncores =</span> <span class="dv">8</span>  )</span></code></pre></div>
</div>
<div id="scfeatures-for-spatial-proteomics" class="section level1">
<h1>scFeatures for spatial proteomics</h1>
<p>The input data is a Seurat object containing <code>celltype</code>,
<code>sample</code>, <code>x_coord</code> and <code>y_coord</code>
label.</p>
<p>Note that, as spatial proteomics contain few genes, the feature
categories of pathway expressions and cell-cell interactions are not
applicable.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># read in data </span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_spatial_proteomics.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;scFeatures&quot;</span>))</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">process_data</span>(data, <span class="at">normalise =</span> F) <span class="co"># here we don&#39;t normalise , because the data has already been normalised </span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="co"># cell type proportions</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>feature_proportion_raw <span class="ot">&lt;-</span> <span class="fu">run_proportion_raw</span>(data, <span class="at">type =</span> <span class="st">&quot;spatial_p&quot;</span>  )</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>feature_proportion_logit <span class="ot">&lt;-</span> <span class="fu">run_proportion_logit</span>(data, <span class="at">type =</span> <span class="st">&quot;spatial_p&quot;</span>)</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>feature_proportion_ratio <span class="ot">&lt;-</span> <span class="fu">run_proportion_ratio</span>(data, <span class="at">type =</span> <span class="st">&quot;spatial_p&quot;</span>)</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="co"># cell type specific gene expressions</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>feature_gene_mean_celltype <span class="ot">&lt;-</span> <span class="fu">run_gene_mean_celltype</span>(data,  <span class="at">type =</span> <span class="st">&quot;spatial_p&quot;</span>  )</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>feature_gene_prop_celltype  <span class="ot">&lt;-</span> <span class="fu">run_gene_prop_celltype</span>( data, <span class="at">type =</span> <span class="st">&quot;spatial_p&quot;</span> )</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>feature_gene_cor_celltype <span class="ot">&lt;-</span> <span class="fu">run_gene_cor_celltype</span>(data, <span class="at">type =</span> <span class="st">&quot;spatial_p&quot;</span> )</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Bulk expressions </span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>feature_gene_mean_bulk <span class="ot">&lt;-</span> <span class="fu">run_gene_mean</span>(data, <span class="at">type =</span> <span class="st">&quot;spatial_p&quot;</span> )</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>feature_gene_prop_bulk <span class="ot">&lt;-</span> <span class="fu">run_gene_prop</span>(data, <span class="at">type =</span> <span class="st">&quot;spatial_p&quot;</span> )</span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>feature_gene_cor_bulk <span class="ot">&lt;-</span> <span class="fu">run_gene_cor</span>(data, <span class="at">type =</span> <span class="st">&quot;spatial_p&quot;</span> )</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Spatial metrics</span></span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>feature_L_stats <span class="ot">&lt;-</span> <span class="fu">run_L_function</span>( data, <span class="at">type =</span> <span class="st">&quot;spatial_p&quot;</span>   )</span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>feature_morans_I <span class="ot">&lt;-</span> <span class="fu">run_Morans_I</span>( data, <span class="at">type =</span> <span class="st">&quot;spatial_p&quot;</span>  )</span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a>feature_celltype_interaction <span class="ot">&lt;-</span> <span class="fu">run_celltype_interaction</span>( data, <span class="at">type =</span> <span class="st">&quot;spatial_p&quot;</span> )</span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a>feature_nn_correlation <span class="ot">&lt;-</span> <span class="fu">run_nn_correlation</span>( data, <span class="at">type =</span> <span class="st">&quot;spatial_p&quot;</span>  )</span></code></pre></div>
</div>
<div id="scfeature-for-spatial-transcriptomics" class="section level1">
<h1>scFeature for spatial transcriptomics</h1>
<p>The input data is a Seurat object containing <code>sample</code>,
<code>x_cord</code> and <code>y_cord</code>.</p>
<p>Additionally, there are two assays needed, the <code>RNA</code>,
which contains the gene expression of each spot and the
<code>predictions</code>.</p>
<p>The <code>predictions</code> assay is a matrix in the form of cell
types x spot, which stores the cell type probability of each spot. This
can be obtained from performing cell type prediction using reference
data.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co"># read in data </span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span>   <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_spatial_transcriptomics.rds&quot;</span>, <span class="at">package =</span> <span class="st">&quot;scFeatures&quot;</span>))</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">process_data</span>(data, <span class="at">normalise=</span>F) <span class="co"># here we don&#39;t normalise , because the data has already been normalised </span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">get_num_cell_per_spot</span>(data) <span class="co"># for spatial transcriptomics, we need to estimate the number of cells in each spot</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="co"># format the patient ID as `a_cond_b` eg, `patient09_cond_nonresponder`. This is so that we can retrieve the sample ID and the corresponding condition when running downstream analysis on the generated features </span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>sample <span class="ot">&lt;-</span> <span class="fu">paste0</span>( data<span class="sc">$</span>sample  , <span class="st">&quot;_cond_&quot;</span>, data<span class="sc">$</span>condition)</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="co"># cell type proportions</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>feature_proportion_raw <span class="ot">&lt;-</span> <span class="fu">run_proportion_raw</span>(data, <span class="at">type =</span>  <span class="st">&quot;spatial_t&quot;</span>  )</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>feature_proportion_logit <span class="ot">&lt;-</span> <span class="fu">run_proportion_logit</span>(data , <span class="at">type =</span>  <span class="st">&quot;spatial_t&quot;</span> )</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>feature_proportion_ratio <span class="ot">&lt;-</span> <span class="fu">run_proportion_ratio</span>(data, <span class="at">type =</span>  <span class="st">&quot;spatial_t&quot;</span> )</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a><span class="co"># cell type specific gene expressions</span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>data_remove_mito <span class="ot">&lt;-</span> <span class="fu">remove_mito</span>(data)</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>feature_gene_mean_celltype <span class="ot">&lt;-</span> <span class="fu">run_gene_mean_celltype</span>( data_remove_mito, <span class="at">type =</span>  <span class="st">&quot;spatial_t&quot;</span> )</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a><span class="co"># cell type specific pathway expressions </span></span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a>feature_pathway_mean <span class="ot">&lt;-</span> <span class="fu">run_pathway_mean</span>(data, <span class="at">geneset =</span> <span class="cn">NULL</span>,</span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">species =</span> <span class="st">&quot;Mus musculus&quot;</span>,</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">type =</span>  <span class="st">&quot;spatial_t&quot;</span>  )</span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Here the data is from mouse samples, therefore we need to set the species accordingly.</span></span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-34"><a href="#cb6-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-35"><a href="#cb6-35" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb6-36"><a href="#cb6-36" aria-hidden="true" tabindex="-1"></a><span class="co"># Bulk expressions </span></span>
<span id="cb6-37"><a href="#cb6-37" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb6-38"><a href="#cb6-38" aria-hidden="true" tabindex="-1"></a>feature_gene_mean_bulk <span class="ot">&lt;-</span> <span class="fu">run_gene_mean</span>(data, <span class="at">type =</span>  <span class="st">&quot;spatial_t&quot;</span> )</span>
<span id="cb6-39"><a href="#cb6-39" aria-hidden="true" tabindex="-1"></a>feature_gene_prop_bulk <span class="ot">&lt;-</span> <span class="fu">run_gene_prop</span>(data, <span class="at">type =</span>  <span class="st">&quot;spatial_t&quot;</span> )</span>
<span id="cb6-40"><a href="#cb6-40" aria-hidden="true" tabindex="-1"></a>feature_gene_cor_bulk <span class="ot">&lt;-</span> <span class="fu">run_gene_cor</span>(data, <span class="at">type =</span>  <span class="st">&quot;spatial_t&quot;</span> )</span>
<span id="cb6-41"><a href="#cb6-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-42"><a href="#cb6-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-43"><a href="#cb6-43" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb6-44"><a href="#cb6-44" aria-hidden="true" tabindex="-1"></a><span class="co"># Spatial metrics</span></span>
<span id="cb6-45"><a href="#cb6-45" aria-hidden="true" tabindex="-1"></a><span class="co">#---------------------------------------------------------------------------</span></span>
<span id="cb6-46"><a href="#cb6-46" aria-hidden="true" tabindex="-1"></a>feature_L_stats <span class="ot">&lt;-</span> <span class="fu">run_L_function</span>( data, <span class="at">type =</span> <span class="st">&quot;spatial_t&quot;</span> )</span>
<span id="cb6-47"><a href="#cb6-47" aria-hidden="true" tabindex="-1"></a>feature_morans_I <span class="ot">&lt;-</span> <span class="fu">run_Morans_I</span>( data, <span class="at">type =</span> <span class="st">&quot;spatial_t&quot;</span> )</span>
<span id="cb6-48"><a href="#cb6-48" aria-hidden="true" tabindex="-1"></a>feature_celltype_interaction <span class="ot">&lt;-</span> <span class="fu">run_celltype_interaction</span>( data, <span class="at">type =</span> <span class="st">&quot;spatial_t&quot;</span>  )</span>
<span id="cb6-49"><a href="#cb6-49" aria-hidden="true" tabindex="-1"></a>feature_nn_correlation <span class="ot">&lt;-</span> <span class="fu">run_nn_correlation</span>( data, <span class="at">type =</span> <span class="st">&quot;spatial_t&quot;</span> )</span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
