<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A summary of scFeatures functions • scFeatures</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="A summary of scFeatures functions">
<meta property="og:description" content="scFeatures">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scFeatures</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scFeatures_associationstudy.html">An association study of features with conditions with scFeatures</a>
    </li>
    <li>
      <a href="../articles/scFeatures_detail.html">A detailed explanation of scFeatures' features</a>
    </li>
    <li>
      <a href="../articles/scFeatures_summary.html">A summary of scFeatures functions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="scFeatures_summary_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A summary of scFeatures functions</h1>
            
      
      
      <div class="hidden name"><code>scFeatures_summary.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette provides a list of all functions in scFeatures for quick reference. For detailed explanation of scFeatures and case studies, please refer to the other vignette “Vignette - detailed explanation of scFeatures”.</p>
<p>If you wish to run downstream analysis using the generated features (see “Vignette - association study”), please make sure that when you run scFeatures, the <code>sample</code> is in the format of <code>a_cond_b</code>. Where <code>a</code> is sample ID and <code>b</code> is the condition, e.g., <code>patient09_cond_nonresponder</code>. This is so that we can retrieve the sample ID and the corresponding condition when running the analysis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scFeatures</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-run-of-scfeatures">Quick run of scFeatures<a class="anchor" aria-label="anchor" href="#quick-run-of-scfeatures"></a>
</h2>
<p>scFeatures can be run using one line of code <code>scfeatures_result &lt;- scFeatures(data)</code>, which generates a list of data frames containing all feature types in the form of samples x features.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_scrnaseq.rds"</span>, package <span class="op">=</span> <span class="st">"scFeatures"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">data</span>, normalise <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co">#  perform normalisation </span></span>
<span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scFeatures.html">scFeatures</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<p>Below we list the function for generating each individual feature types for scRNA-seq, spatial proteomic and spatial transcriptomic data.</p>
</div>
<div class="section level2">
<h2 id="scfeatures-for-single-cell-rna-seq-data">scFeatures for single-cell RNA-seq data<a class="anchor" aria-label="anchor" href="#scfeatures-for-single-cell-rna-seq-data"></a>
</h2>
<p>scFeatures accept a Seurat object containing the <code>sample</code> and <code>celltype</code> columns.<br>
For SingleCellExperiment or SpatialExperiment object, we provide a function <code><a href="../reference/makeSeurat.html">makeSeurat()</a></code> for the conversion.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># read in data </span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_scrnaseq.rds"</span>, package <span class="op">=</span> <span class="st">"scFeatures"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">data</span>, normalise <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co">#  perform normalisation </span></span>
<span></span>
<span><span class="co"># format the patient ID as `a_cond_b` eg, `patient09_cond_nonresponder`. This is so that we can retrieve the sample ID and the corresponding condition when running downstream analysis on the generated features </span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span> <span class="va">data</span><span class="op">$</span><span class="va">sample</span>  , <span class="st">"_cond_"</span>, <span class="va">data</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type proportions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_proportion_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_raw.html">run_proportion_raw</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">feature_proportion_logit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_logit.html">run_proportion_logit</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">feature_proportion_ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_ratio.html">run_proportion_ratio</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type specific gene expressions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">data_remove_mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_mito.html">remove_mito</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="co"># optional step, run if mitochondria amd ribosomal genes are not of interest </span></span>
<span></span>
<span><span class="co"># by default, use the top variable genes</span></span>
<span><span class="co"># users can change the number of top variable genes by changing `num_top_gene` </span></span>
<span><span class="co"># alternatively, users can also input their genes of interest in a dataframe format</span></span>
<span><span class="va">genes_of_interest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> marker <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"S100A11"</span> , <span class="st">"GZMB"</span> ,  <span class="st">"DUSP1"</span> <span class="op">)</span> , </span>
<span>                                celltype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"CD8, T Effector Memory"</span> , <span class="st">"CD8, T Effector Memory"</span>  ,  <span class="st">"Naive T Cells"</span><span class="op">)</span> <span class="op">)</span> </span>
<span></span>
<span><span class="va">feature_gene_mean_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean_celltype.html">run_gene_mean_celltype</a></span><span class="op">(</span><span class="va">data_remove_mito</span><span class="op">)</span>  </span>
<span><span class="va">feature_gene_prop_celltype</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_prop_celltype.html">run_gene_prop_celltype</a></span><span class="op">(</span><span class="va">data_remove_mito</span>, genes <span class="op">=</span> <span class="va">genes_of_interest</span><span class="op">)</span></span>
<span><span class="va">feature_gene_cor_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_cor_celltype.html">run_gene_cor_celltype</a></span><span class="op">(</span><span class="va">data_remove_mito</span>,  num_top_gene <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span><span class="co"># by default, pick around 100 most variable genes per cell type, can change this number through the num_top_gene argument </span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type specific pathway expressions </span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_pathway_gsva</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathway_gsva.html">run_pathway_gsva</a></span><span class="op">(</span><span class="va">data</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span><span class="op">)</span> </span>
<span><span class="va">feature_pathway_mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathway_mean.html">run_pathway_mean</a></span><span class="op">(</span><span class="va">data</span> , species <span class="op">=</span> <span class="st">"Homo sapiens"</span><span class="op">)</span></span>
<span><span class="co"># by default, use the 50 hallmark pathways, users can also input their gene set of interest in a list format </span></span>
<span><span class="va">geneset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"pathway_a"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CAPNS1"</span>, <span class="st">"TLCD1"</span><span class="op">)</span>, </span>
<span>                <span class="st">"pathway_b"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PEX6"</span>,<span class="st">"DPRXP4"</span> <span class="op">)</span><span class="op">)</span>  </span>
<span><span class="va">feature_pathway_prop</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathway_prop.html">run_pathway_prop</a></span><span class="op">(</span><span class="va">data</span>, geneset<span class="op">=</span><span class="va">geneset</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type specific cell-cell interactions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_CCI</span> <span class="op">&lt;-</span> <span class="fu">run_CCI</span><span class="op">(</span><span class="va">data</span> , species <span class="op">=</span> <span class="st">"Homo sapiens"</span> <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># Bulk expressions </span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span></span>
<span><span class="co"># by default, use the top variable genes</span></span>
<span><span class="co"># users can change the number of top variable genes by changing `num_top_gene` </span></span>
<span><span class="co"># alternatively, users can also input their genes of interest in a list format</span></span>
<span><span class="va">feature_gene_mean_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean.html">run_gene_mean</a></span><span class="op">(</span><span class="va">data</span> <span class="op">)</span></span>
<span><span class="va">feature_gene_cor_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_cor.html">run_gene_cor</a></span><span class="op">(</span><span class="va">data</span> ,  num_top_gene <span class="op">=</span> <span class="fl">5</span> <span class="op">)</span></span>
<span><span class="va">genes_of_interest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TIGIT"</span>, <span class="st">"PDCD1"</span><span class="op">)</span></span>
<span><span class="va">feature_gene_prop_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_prop.html">run_gene_prop</a></span><span class="op">(</span><span class="va">data</span>, genes <span class="op">=</span> <span class="va">genes_of_interest</span> <span class="op">)</span></span>
<span><span class="co"># by default, pick 1500 most variable genes, can change this number through the num_top_gene argument </span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="scfeatures-for-spatial-proteomics">scFeatures for spatial proteomics<a class="anchor" aria-label="anchor" href="#scfeatures-for-spatial-proteomics"></a>
</h2>
<p>The input data is a Seurat object containing <code>celltype</code>, <code>sample</code>, <code>x_coord</code> and <code>y_coord</code> columns.</p>
<p>For SingleCellExperiment or SpatialExperiment object, we provide a function <code><a href="../reference/makeSeurat.html">makeSeurat()</a></code> for the conversion.</p>
<p>Note that, as spatial proteomics contain few genes, the feature categories of pathway expressions and cell-cell interactions are not applicable.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># read in data </span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_spatial_proteomics.rds"</span>, package <span class="op">=</span> <span class="st">"scFeatures"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">data</span>, normalise <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># here we don't normalise , because the data has already been normalised </span></span>
<span> </span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type proportions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_proportion_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_raw.html">run_proportion_raw</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span>  <span class="op">)</span></span>
<span><span class="va">feature_proportion_logit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_logit.html">run_proportion_logit</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_proportion_ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_ratio.html">run_proportion_ratio</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type specific gene expressions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_gene_mean_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean_celltype.html">run_gene_mean_celltype</a></span><span class="op">(</span><span class="va">data</span>,  type <span class="op">=</span> <span class="st">"spatial_p"</span>  <span class="op">)</span></span>
<span><span class="va">feature_gene_prop_celltype</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_prop_celltype.html">run_gene_prop_celltype</a></span><span class="op">(</span> <span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span> <span class="op">)</span></span>
<span><span class="va">feature_gene_cor_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_cor_celltype.html">run_gene_cor_celltype</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span> <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># Bulk expressions </span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_gene_mean_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean.html">run_gene_mean</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span> <span class="op">)</span></span>
<span><span class="va">feature_gene_prop_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_prop.html">run_gene_prop</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span> <span class="op">)</span></span>
<span><span class="va">feature_gene_cor_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_cor.html">run_gene_cor</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span> <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># Spatial metrics</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_L_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_L_function.html">run_L_function</a></span><span class="op">(</span> <span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span>   <span class="op">)</span></span>
<span><span class="va">feature_morans_I</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_Morans_I.html">run_Morans_I</a></span><span class="op">(</span> <span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span>  <span class="op">)</span></span>
<span><span class="va">feature_celltype_interaction</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_celltype_interaction.html">run_celltype_interaction</a></span><span class="op">(</span> <span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span> <span class="op">)</span></span>
<span><span class="va">feature_nn_correlation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_nn_correlation.html">run_nn_correlation</a></span><span class="op">(</span> <span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="scfeature-for-spatial-transcriptomics">scFeature for spatial transcriptomics<a class="anchor" aria-label="anchor" href="#scfeature-for-spatial-transcriptomics"></a>
</h2>
<p>The input data is a Seurat object containing <code>sample</code>, <code>x_cord</code> and <code>y_cord</code> columns.<br>
Additionally, apart from the <code>RNA</code> assay, which contains the gene expression of each spot, a <code>predictions</code> assay is also needed.</p>
<p>The <code>predictions</code> assay is a matrix in the form of cell types x spot, which stores the cell type probability of each spot. This can be obtained from performing cell type prediction using reference data.</p>
<p>For SingleCellExperiment or SpatialExperiment object, we provide a function <code><a href="../reference/makeSeurat.html">makeSeurat()</a></code> for the conversion.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># read in data </span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span>   <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_spatial_transcriptomics.rds"</span>, package <span class="op">=</span> <span class="st">"scFeatures"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">data</span>, normalise<span class="op">=</span><span class="cn">F</span><span class="op">)</span> <span class="co"># here we don't normalise , because the data has already been normalised </span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_num_cell_per_spot.html">get_num_cell_per_spot</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="co"># for spatial transcriptomics, we need to estimate the number of cells in each spot</span></span>
<span></span>
<span><span class="co"># format the patient ID as `a_cond_b` eg, `patient09_cond_nonresponder`. This is so that we can retrieve the sample ID and the corresponding condition when running downstream analysis on the generated features </span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span> <span class="va">data</span><span class="op">$</span><span class="va">sample</span>  , <span class="st">"_cond_"</span>, <span class="va">data</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type proportions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_proportion_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_raw.html">run_proportion_raw</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span>  <span class="st">"spatial_t"</span>  <span class="op">)</span></span>
<span><span class="va">feature_proportion_logit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_logit.html">run_proportion_logit</a></span><span class="op">(</span><span class="va">data</span> , type <span class="op">=</span>  <span class="st">"spatial_t"</span> <span class="op">)</span></span>
<span><span class="va">feature_proportion_ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_ratio.html">run_proportion_ratio</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span>  <span class="st">"spatial_t"</span> <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type specific gene expressions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">data_remove_mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_mito.html">remove_mito</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">feature_gene_mean_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean_celltype.html">run_gene_mean_celltype</a></span><span class="op">(</span> <span class="va">data_remove_mito</span>, type <span class="op">=</span>  <span class="st">"spatial_t"</span> <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type specific pathway expressions </span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_pathway_mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathway_mean.html">run_pathway_mean</a></span><span class="op">(</span><span class="va">data</span>, geneset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                                           species <span class="op">=</span> <span class="st">"Mus musculus"</span>,</span>
<span>                                           type <span class="op">=</span>  <span class="st">"spatial_t"</span>  <span class="op">)</span></span>
<span><span class="co"># Here the data is from mouse samples, therefore we need to set the species accordingly.</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># Bulk expressions </span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_gene_mean_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean.html">run_gene_mean</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span>  <span class="st">"spatial_t"</span> <span class="op">)</span></span>
<span><span class="va">feature_gene_prop_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_prop.html">run_gene_prop</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span>  <span class="st">"spatial_t"</span> <span class="op">)</span></span>
<span><span class="va">feature_gene_cor_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_cor.html">run_gene_cor</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span>  <span class="st">"spatial_t"</span> <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># Spatial metrics</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_L_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_L_function.html">run_L_function</a></span><span class="op">(</span> <span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span> <span class="op">)</span></span>
<span><span class="va">feature_morans_I</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_Morans_I.html">run_Morans_I</a></span><span class="op">(</span> <span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span> <span class="op">)</span></span>
<span><span class="va">feature_celltype_interaction</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_celltype_interaction.html">run_celltype_interaction</a></span><span class="op">(</span> <span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span>  <span class="op">)</span></span>
<span><span class="va">feature_nn_correlation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_nn_correlation.html">run_nn_correlation</a></span><span class="op">(</span> <span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span> <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="sessioninfo">sessionInfo()<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Debian GNU/Linux 11 (bullseye)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.13.so</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] scFeatures_0.99.9      RhpcBLASctl_0.21-247.1 inline_0.3.19         </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] rappdirs_0.3.3              rtracklayer_1.58.0         </span></span>
<span><span class="co">#&gt;   [3] scattermore_0.8             SpatialExperiment_1.8.0    </span></span>
<span><span class="co">#&gt;   [5] R.methodsS3_1.8.2           SeuratObject_4.1.3         </span></span>
<span><span class="co">#&gt;   [7] ragg_1.2.4                  tidyr_1.2.1                </span></span>
<span><span class="co">#&gt;   [9] ggplot2_3.4.0               bit64_4.0.5                </span></span>
<span><span class="co">#&gt;  [11] knitr_1.41                  irlba_2.3.5.1              </span></span>
<span><span class="co">#&gt;  [13] DelayedArray_0.24.0         R.utils_2.12.2             </span></span>
<span><span class="co">#&gt;  [15] data.table_1.14.4           KEGGREST_1.38.0            </span></span>
<span><span class="co">#&gt;  [17] RCurl_1.98-1.9              AnnotationFilter_1.22.0    </span></span>
<span><span class="co">#&gt;  [19] generics_0.1.3              BiocGenerics_0.44.0        </span></span>
<span><span class="co">#&gt;  [21] GenomicFeatures_1.50.2      ScaledMatrix_1.6.0         </span></span>
<span><span class="co">#&gt;  [23] cowplot_1.1.1               RSQLite_2.2.19             </span></span>
<span><span class="co">#&gt;  [25] EnsDb.Mmusculus.v79_2.99.0  RANN_2.6.1                 </span></span>
<span><span class="co">#&gt;  [27] future_1.29.0               bit_4.0.5                  </span></span>
<span><span class="co">#&gt;  [29] spatstat.data_3.0-0         xml2_1.3.3                 </span></span>
<span><span class="co">#&gt;  [31] httpuv_1.6.6                SummarizedExperiment_1.28.0</span></span>
<span><span class="co">#&gt;  [33] assertthat_0.2.1            xfun_0.35                  </span></span>
<span><span class="co">#&gt;  [35] hms_1.1.2                   jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt;  [37] babelgene_22.9              evaluate_0.18              </span></span>
<span><span class="co">#&gt;  [39] promises_1.2.0.1            fansi_1.0.3                </span></span>
<span><span class="co">#&gt;  [41] restfulr_0.0.15             progress_1.2.2             </span></span>
<span><span class="co">#&gt;  [43] dbplyr_2.2.1                igraph_1.3.5               </span></span>
<span><span class="co">#&gt;  [45] DBI_1.1.3                   htmlwidgets_1.5.4          </span></span>
<span><span class="co">#&gt;  [47] spatstat.geom_3.0-3         stats4_4.2.1               </span></span>
<span><span class="co">#&gt;  [49] purrr_0.3.5                 ellipsis_0.3.2             </span></span>
<span><span class="co">#&gt;  [51] dplyr_1.0.10                annotate_1.76.0            </span></span>
<span><span class="co">#&gt;  [53] RcppParallel_5.1.5          biomaRt_2.54.0             </span></span>
<span><span class="co">#&gt;  [55] deldir_1.0-6                sparseMatrixStats_1.10.0   </span></span>
<span><span class="co">#&gt;  [57] MatrixGenerics_1.10.0       vctrs_0.5.1                </span></span>
<span><span class="co">#&gt;  [59] SingleCellExperiment_1.20.0 Biobase_2.58.0             </span></span>
<span><span class="co">#&gt;  [61] ensembldb_2.22.0            ROCR_1.0-11                </span></span>
<span><span class="co">#&gt;  [63] abind_1.4-5                 cachem_1.0.6               </span></span>
<span><span class="co">#&gt;  [65] progressr_0.11.0            sctransform_0.3.5          </span></span>
<span><span class="co">#&gt;  [67] GenomicAlignments_1.34.0    prettyunits_1.1.1          </span></span>
<span><span class="co">#&gt;  [69] goftest_1.2-3               cluster_2.1.4              </span></span>
<span><span class="co">#&gt;  [71] ape_5.6-2                   lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt;  [73] crayon_1.5.2                spatstat.explore_3.0-5     </span></span>
<span><span class="co">#&gt;  [75] edgeR_3.40.0                pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [77] GenomeInfoDb_1.34.4         nlme_3.1-160               </span></span>
<span><span class="co">#&gt;  [79] ProtGenerics_1.30.0         rlang_1.0.6                </span></span>
<span><span class="co">#&gt;  [81] globals_0.16.2              lifecycle_1.0.3            </span></span>
<span><span class="co">#&gt;  [83] miniUI_0.1.1.1              filelock_1.0.2             </span></span>
<span><span class="co">#&gt;  [85] BiocFileCache_2.6.0         rsvd_1.0.5                 </span></span>
<span><span class="co">#&gt;  [87] rprojroot_2.0.3             polyclip_1.10-4            </span></span>
<span><span class="co">#&gt;  [89] GSVA_1.46.0                 matrixStats_0.63.0         </span></span>
<span><span class="co">#&gt;  [91] lmtest_0.9-40               graph_1.76.0               </span></span>
<span><span class="co">#&gt;  [93] Matrix_1.5-3                Rhdf5lib_1.20.0            </span></span>
<span><span class="co">#&gt;  [95] zoo_1.8-11                  ggridges_0.5.4             </span></span>
<span><span class="co">#&gt;  [97] png_0.1-8                   viridisLite_0.4.1          </span></span>
<span><span class="co">#&gt;  [99] rjson_0.2.21                bitops_1.0-7               </span></span>
<span><span class="co">#&gt; [101] R.oo_1.25.0                 KernSmooth_2.23-20         </span></span>
<span><span class="co">#&gt; [103] rhdf5filters_1.10.0         Biostrings_2.66.0          </span></span>
<span><span class="co">#&gt; [105] blob_1.2.3                  DelayedMatrixStats_1.20.0  </span></span>
<span><span class="co">#&gt; [107] stringr_1.5.0               parallelly_1.32.1          </span></span>
<span><span class="co">#&gt; [109] spatstat.random_3.0-1       S4Vectors_0.36.0           </span></span>
<span><span class="co">#&gt; [111] beachmat_2.14.0             scales_1.2.1               </span></span>
<span><span class="co">#&gt; [113] memoise_2.0.1               GSEABase_1.60.0            </span></span>
<span><span class="co">#&gt; [115] magrittr_2.0.3              plyr_1.8.8                 </span></span>
<span><span class="co">#&gt; [117] ica_1.0-3                   zlibbioc_1.44.0            </span></span>
<span><span class="co">#&gt; [119] compiler_4.2.1              dqrng_0.3.0                </span></span>
<span><span class="co">#&gt; [121] BiocIO_1.8.0                RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt; [123] fitdistrplus_1.1-8          Rsamtools_2.14.0           </span></span>
<span><span class="co">#&gt; [125] cli_3.4.1                   XVector_0.38.0             </span></span>
<span><span class="co">#&gt; [127] listenv_0.8.0               EnsDb.Hsapiens.v79_2.99.0  </span></span>
<span><span class="co">#&gt; [129] patchwork_1.1.2             pbapply_1.6-0              </span></span>
<span><span class="co">#&gt; [131] MASS_7.3-58.1               tidyselect_1.2.0           </span></span>
<span><span class="co">#&gt; [133] stringi_1.7.8               textshaping_0.3.6          </span></span>
<span><span class="co">#&gt; [135] yaml_2.3.6                  BiocSingular_1.14.0        </span></span>
<span><span class="co">#&gt; [137] locfit_1.5-9.6              ggrepel_0.9.2              </span></span>
<span><span class="co">#&gt; [139] grid_4.2.1                  sass_0.4.3                 </span></span>
<span><span class="co">#&gt; [141] tools_4.2.1                 future.apply_1.10.0        </span></span>
<span><span class="co">#&gt; [143] parallel_4.2.1              AUCell_1.20.1              </span></span>
<span><span class="co">#&gt; [145] gridExtra_2.3               Rtsne_0.16                 </span></span>
<span><span class="co">#&gt; [147] DropletUtils_1.18.0         proxyC_0.3.3               </span></span>
<span><span class="co">#&gt; [149] digest_0.6.30               shiny_1.7.3                </span></span>
<span><span class="co">#&gt; [151] Rcpp_1.0.9                  GenomicRanges_1.50.1       </span></span>
<span><span class="co">#&gt; [153] scuttle_1.8.0               later_1.3.0                </span></span>
<span><span class="co">#&gt; [155] RcppAnnoy_0.0.20            httr_1.4.4                 </span></span>
<span><span class="co">#&gt; [157] AnnotationDbi_1.60.0        colorspace_2.0-3           </span></span>
<span><span class="co">#&gt; [159] XML_3.99-0.13               fs_1.5.2                   </span></span>
<span><span class="co">#&gt; [161] tensor_1.5                  reticulate_1.26            </span></span>
<span><span class="co">#&gt; [163] IRanges_2.32.0              splines_4.2.1              </span></span>
<span><span class="co">#&gt; [165] uwot_0.1.14                 spatstat.utils_3.0-1       </span></span>
<span><span class="co">#&gt; [167] pkgdown_2.0.6               sp_1.5-1                   </span></span>
<span><span class="co">#&gt; [169] plotly_4.10.1               systemfonts_1.0.4          </span></span>
<span><span class="co">#&gt; [171] xtable_1.8-4                jsonlite_1.8.3             </span></span>
<span><span class="co">#&gt; [173] R6_2.5.1                    pillar_1.8.1               </span></span>
<span><span class="co">#&gt; [175] htmltools_0.5.3             mime_0.12                  </span></span>
<span><span class="co">#&gt; [177] glue_1.6.2                  fastmap_1.1.0              </span></span>
<span><span class="co">#&gt; [179] BiocParallel_1.32.4         codetools_0.2-18           </span></span>
<span><span class="co">#&gt; [181] utf8_1.2.2                  lattice_0.20-45            </span></span>
<span><span class="co">#&gt; [183] bslib_0.4.1                 spatstat.sparse_3.0-0      </span></span>
<span><span class="co">#&gt; [185] tibble_3.1.8                curl_4.3.3                 </span></span>
<span><span class="co">#&gt; [187] leiden_0.4.3                gtools_3.9.3               </span></span>
<span><span class="co">#&gt; [189] magick_2.7.3                survival_3.4-0             </span></span>
<span><span class="co">#&gt; [191] limma_3.54.0                rmarkdown_2.18             </span></span>
<span><span class="co">#&gt; [193] desc_1.4.2                  munsell_0.5.0              </span></span>
<span><span class="co">#&gt; [195] rhdf5_2.42.0                GenomeInfoDbData_1.2.9     </span></span>
<span><span class="co">#&gt; [197] HDF5Array_1.26.0            reshape2_1.4.4             </span></span>
<span><span class="co">#&gt; [199] gtable_0.3.1                msigdbr_7.5.1              </span></span>
<span><span class="co">#&gt; [201] Seurat_4.3.0</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yue Cao, Yingxin Lin, Ellis Patrick, Pengyi Yang, Jean Yee Hwa Yang.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
