<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scFeatures">
<title>A detailed explanation of scFeatures' features • scFeatures</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="A detailed explanation of scFeatures' features">
<meta property="og:description" content="scFeatures">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scFeatures</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.9</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/scFeatures_associationstudy.html">An association study of features with conditions with scFeatures</a>
    <a class="dropdown-item" href="../articles/scFeatures_detail.html">A detailed explanation of scFeatures' features</a>
    <a class="dropdown-item" href="../articles/scFeatures_summary.html">An overview of scFeatures' functions</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/SydneyBioX/scFeatures/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>A detailed explanation of scFeatures' features</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/SydneyBioX/scFeatures/blob/HEAD/vignettes/scFeatures_detail.Rmd" class="external-link"><code>vignettes/scFeatures_detail.Rmd</code></a></small>
      <div class="d-none name"><code>scFeatures_detail.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette provides a detailed description of scFeatures,
including description for each feature category, the expected output and
case studies of using the generated features for classification task and
survival analysis.</p>
<p>For a quick starting point (the TLDR version), please see the other
vignette “Vignette - summary of scFeatures functions”.</p>
<p>Note that all scFeatures functions contain a <code>ncores</code>
argument that controls the number of cores. The default is 1 core.</p>
<p>If you wish to run downstream analysis using the generated features
(see “Vignette - association study”), please make sure that when you run
scFeatures, the <code>sample</code> is in the format of
<code>a_cond_b</code>. Where <code>a</code> is sample ID and
<code>b</code> is the condition, eg,
<code>patient09_cond_nonresponder</code>. This is so that we can
retrieve the sample ID and the corresponding condition when running the
analysis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scFeatures</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-run-of-scfeatures">Quick run of scFeatures<a class="anchor" aria-label="anchor" href="#quick-run-of-scfeatures"></a>
</h2>
<p>scFeatures can be run using one line of code
<code>scfeatures_result &lt;- scFeatures(data)</code>, which generates a
list of dataframes containing all feature types in the form of samples x
features.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_scrnaseq.rds"</span>, package <span class="op">=</span> <span class="st">"scFeatures"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">data</span>, normalise <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co">#  perform normalisation</span></span>
<span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scFeatures.html">scFeatures</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "generating CCI features"</span></span></code></pre></div>
<p>The above function uses all default settings and generate all feature
types. We provide an example below that describes how this can be done
with a spatial proteomics dataset.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># read in data</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_scrnaseq.rds"</span>, package <span class="op">=</span> <span class="st">"scFeatures"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># randomly generate some x and y coordinate to make this a </span></span>
<span><span class="co"># toy "spatial proteomics" data</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> , replace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> , replace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSeurat.html">makeSeurat</a></span><span class="op">(</span><span class="va">data</span>, spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">data</span>, normalise <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>  </span>
<span></span>
<span></span>
<span></span>
<span><span class="co"># select feature types to compute.</span></span>
<span><span class="va">feature_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"proportion_raw"</span>, <span class="st">"pathway_gsva"</span>, <span class="st">"L_stats"</span>, <span class="st">"gene_mean_celltype"</span>,</span>
<span>    <span class="st">"gene_prop_aggregated"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scFeatures.html">scFeatures</a></span><span class="op">(</span><span class="va">data</span>,</span>
<span>    feature_types <span class="op">=</span> <span class="va">feature_types</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"spatial_p"</span>,</span>
<span>    <span class="co"># set to user specified pathways.</span></span>
<span>    geneset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        <span class="st">"pathway_a"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span>,</span>
<span>        <span class="st">"pathway_b"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">[</span><span class="fl">51</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="co"># by default uses top variable genes to generate the celltype</span></span>
<span>    <span class="co"># specific gene expression feature, now set to user defined genes.</span></span>
<span>    celltype_genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>        celltype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Naive T Cells"</span> , <span class="fl">10</span><span class="op">)</span>,  <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span> <span class="st">"Cytotoxic CD8"</span> , <span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        marker <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">[</span><span class="fl">11</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="co"># by default uses top variable genes to generate the overall</span></span>
<span>    <span class="co"># aggregated gene expression feature category, now set to user</span></span>
<span>    <span class="co"># defined genes.</span></span>
<span>    aggregated_genes <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span>,</span>
<span>    <span class="co"># by default uses single core</span></span>
<span>    ncores <span class="op">=</span> <span class="fl">8</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="scfeatures-on-single-cell-rna-seq-data">scFeatures on single-cell RNA-seq data<a class="anchor" aria-label="anchor" href="#scfeatures-on-single-cell-rna-seq-data"></a>
</h2>
<p>For demonstration purpose, we provide a subsampled version of the
melanoma pre-treatment dataset [1] as used in our manuscript.</p>
<p>scFeatures takes in data as a Seurat object, with the gene expression
stored in the assay <code>RNA</code>. The required metadata is the
<code>celltype</code> and <code>sample</code> label. The
<code>condition</code> label is optional, and is only required if users
wish to learn the difference of disease condition using the generated
features.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_scrnaseq.rds"</span>, package <span class="op">=</span> <span class="st">"scFeatures"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="co"># some pre-processing, including normalisation</span></span>
<span></span>
<span><span class="co"># confirm the data contain cell types</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Naive T Cells"          "Cytotoxic CD8"          "CD8, T Effector Memory"</span></span>
<span></span>
<span><span class="co"># confirm the data contain samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "Pre_P8"  "Pre_P27" "Pre_P7"  "Pre_P20" "Pre_P3"</span></span>
<span></span>
<span><span class="co"># disease conditions in this data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Responder"     "Non-responder"</span></span>
<span></span>
<span><span class="co"># format the patient ID as `a_cond_b` eg, `patient09_cond_nonresponder`. This</span></span>
<span><span class="co"># is so that we can retrieve the sample ID and the corresponding condition when</span></span>
<span><span class="co"># running downstream analysis on the generated features</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">sample</span>, <span class="st">"_cond_"</span>, <span class="va">data</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "Pre_P8_cond_Responder"      "Pre_P27_cond_Non-responder"</span></span>
<span><span class="co">#&gt; [3] "Pre_P7_cond_Responder"      "Pre_P20_cond_Non-responder"</span></span>
<span><span class="co">#&gt; [5] "Pre_P3_cond_Non-responder"</span></span></code></pre></div>
<div class="section level3">
<h3 id="feature-generation">Feature generation<a class="anchor" aria-label="anchor" href="#feature-generation"></a>
</h3>
<p>Here we demonstrate the usage of scFeatures to extract features from
scRNA-seq data. Note that, for the <code>type</code> argument, the
currently supported options are:<br>
* <code>scrna</code> : which stands for single-cell RNA-seq data *
<code>spatial_p</code>: which stands for spatial proteomics data<br>
* <code>spatial_t</code>: which stands for spatial transcriptomics
data</p>
<p>By default, the type is set to <code>scrna</code>, therefore it is
not necessarily to specify the type when the input data is scRNA-seq
data.</p>
<div class="section level4">
<h4 id="cell-type-proportions">Cell type proportions<a class="anchor" aria-label="anchor" href="#cell-type-proportions"></a>
</h4>
<ul>
<li>Detail:<br>
For each sample, we calculate the proportion of each cell type:
<ul>
<li>
<code>Proportion raw</code>: the untransformed proportion.<br>
</li>
<li>
<code>Proportion logit</code>: the logit-transformed proportion
value, which is a common transformation used for proportional
data.<br>
</li>
<li>
<code>Proportion ratio</code>: calculates the ratio of proportion
between two cell types.</li>
</ul>
</li>
<li>Expected output:
<ul>
<li>
<code>Proportion raw</code> and <code>Proportion logit</code>: the
features are in the form of celltype a, celltype b, with the number
representing proportions.</li>
<li>
<code>Proportion ratio</code>: the features are in the form of
celltype a vs celltype b, celltype a vs celltype c, with the number
representing the ratio between the two cell types.</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_proportion_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_raw.html">run_proportion_raw</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">feature_proportion_logit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_logit.html">run_proportion_logit</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">feature_proportion_ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_ratio.html">run_proportion_ratio</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="cell-type-specific-gene-expressions">Cell type specific gene expressions<a class="anchor" aria-label="anchor" href="#cell-type-specific-gene-expressions"></a>
</h4>
<ul>
<li>
<p>Detail:<br>
This feature category aggregate the cells across each cell type and then
construct various features. By default, (if the <code>features</code>
argument is not provided), we restrict to the top variable genes to
reduce the dimensions of the feature. The argument
<code>num_top_gene</code> determines how many genes to include.</p>
<ul>
<li>
<code>Gene mean celltype</code>: the cell type specific gene
expression.</li>
<li>
<code>Gene proportion celltype</code>: for a particular gene, the
percentage of cells in a cell type that it is expressed.</li>
<li>
<code>Gene correlation celltype</code>: correlation of expressions
of two given genes in a cell type.</li>
</ul>
</li>
</ul>
<p>Alternatively, users can provide their own gene of interest to the
<code>features</code> argument.</p>
<ul>
<li>Expected output:
<ul>
<li>
<code>Gene mean celltype</code>: the features are in the form of
gene 1 celltype a, gene 2 celltype b … etc, with the number representing
average gene expression of the given gene across the cells of the the
given celltype.</li>
<li>
<code>Gene proportion celltype</code>: the features are in the form
of gene 1 celltype a, gene 2 celltype b … etc, with the number
representing average gene expression of the given gene across the cells
of the the given celltype.</li>
<li>
<code>Gene correlation celltype</code>: the features are in the form
of gene 1 vs gene 2 cell type a , gene 1 vs gene 3 cell type b … etc,
with the numbers representing the correlation of the two given genes in
the given cell type.</li>
</ul>
</li>
</ul>
<p>Note that mitochondria and ribosomal genes may not be as interesting
to look at. Therefore, users can choose to remove these genes using the
function <code>remove_mito</code>. This step is optional.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_remove_mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_mito.html">remove_mito</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="va">feature_gene_mean_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean_celltype.html">run_gene_mean_celltype</a></span><span class="op">(</span><span class="va">data_remove_mito</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># we can provide our own sets of genes</span></span>
<span><span class="va">genes_of_interest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    marker <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"S100A11"</span>, <span class="st">"GZMB"</span>, <span class="st">"DUSP1"</span><span class="op">)</span>,</span>
<span>    celltype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"CD8, T Effector Memory"</span>,</span>
<span>        <span class="st">"CD8, T Effector Memory"</span>,</span>
<span>        <span class="st">"Naive T Cells"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">feature_gene_prop_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_prop_celltype.html">run_gene_prop_celltype</a></span><span class="op">(</span></span>
<span>    <span class="va">data_remove_mito</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes_of_interest</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># we can change the number to pick ~10 top variables genes per cell type</span></span>
<span><span class="va">feature_gene_cor_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_cor_celltype.html">run_gene_cor_celltype</a></span><span class="op">(</span></span>
<span>    <span class="va">data_remove_mito</span>,</span>
<span>    num_top_gene <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="cell-type-specific-pathway-expressions">Cell type specific pathway expressions<a class="anchor" aria-label="anchor" href="#cell-type-specific-pathway-expressions"></a>
</h4>
<ul>
<li>Detail:<br>
This feature category constructs features relating to pathway (gene set)
expression. By default (when the <code>geneset</code> argument is not
specified), we use the 50 hallmark gene set from msigdb. The users can
also provide their geneset of interest in a list format, with each list
entry containing a vector of the names of genes in a gene set.
<ul>
<li>
<code>pathway_gsva</code>: uses the GSVA function to calculate the
gene set enrichment score for individual cells, then aggregates the
scores across cells within a cell type.</li>
<li>
<code>pathway_mean</code>: averages the gene expression of genes in
a pathway.</li>
<li>
<code>pathway_prop</code>: for the genes in a pathway, calculates
the percentage of cells that these genes are expressed in each cell
type.</li>
</ul>
</li>
<li>Expected output:
<ul>
<li>
<code>pathway_gsva</code> the features are in the form of pathway 1
celltype a, pathway 2 celltype b … etc, with the number representing the
gene set enrichment score of a given pathway in cells from a given
celltype.<br>
</li>
<li>
<code>pathway_mean</code> the features are in the form of pathway 1
celltype a, pathway 2 celltype b … etc, with the number representing the
averaged expression of a given pathway in cells from a given
celltype.<br>
</li>
<li>
<code>pathway_prop</code> the features are in the form of pathway 1
celltype a, pathway 2 celltype b … etc, with the number representing the
proportion of expression of a given pathway in cells from a given
celltype.</li>
</ul>
</li>
</ul>
<p>Note that it is necessarily to indicate whether the species is “Homo
sapiens” or “Mus musculus”. The default is “Homo sapiens”.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This function can take a while. Therefore for larger datasets (eg, over</span></span>
<span><span class="co"># 30,000 cells), we provide a subsample function to increase the speed.</span></span>
<span><span class="va">feature_pathway_gsva</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathway_gsva.html">run_pathway_gsva</a></span><span class="op">(</span></span>
<span>    <span class="va">data</span>,</span>
<span>    geneset <span class="op">=</span> <span class="cn">NULL</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, subsample <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">feature_pathway_mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathway_mean.html">run_pathway_mean</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">feature_pathway_prop</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathway_prop.html">run_pathway_prop</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="cell-type-specific-cell-cell-communications">Cell type specific cell-cell communications<a class="anchor" aria-label="anchor" href="#cell-type-specific-cell-cell-communications"></a>
</h4>
<ul>
<li>Detail:
<ul>
<li>
<code>CCI</code>: This uses the <code>SingleCellSignalR</code>
package to calculate the cell-cell interaction score of ligand recepor
pairs.</li>
</ul>
</li>
<li>Expected output:
<ul>
<li>
<code>CCI</code>: The features are in the form of ligand 1 receptor
2 celltype a, ligand 1 receptor 2 celltype b … etc, with the numbers
representing cell-cell interaction probability.</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_CCI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_CCI.html">run_CCI</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level4">
<h4 id="overall-aggregated-gene-expressions">Overall aggregated gene expressions<a class="anchor" aria-label="anchor" href="#overall-aggregated-gene-expressions"></a>
</h4>
<ul>
<li><p>Detail: The feature types in this category is calculated based on
the aggregated expressions across all cells. The argument
<code>num_top_gene</code> determines how many genes to include. By
default (if the <code>features</code> argument is not provided), the
algorithm picks the top variable genes.</p></li>
<li>
<p>Expected output:</p>
<ul>
<li>
<code>Gene mean</code>: The features are in the form of gene 1, gene
2 … etc, with the numbers representing averaged gene expression across
all cells.<br>
</li>
<li>
<code>Gene cor</code>: The features are in the form of gene 1 vs
gene 2, gene 1 vs gene 3 … etc, with the numbers representing
correlation of gene expressions.<br>
</li>
<li>
<code>Gene prop</code>: The features are in the form of gene 1, gene
2 … etc, with the numbers representing the proportion that the gene is
expressed across all cells.</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_gene_mean_aggregated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean.html">run_gene_mean</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes_of_interest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TIGIT"</span>, <span class="st">"PDCD1"</span><span class="op">)</span></span>
<span><span class="va">feature_gene_cor_aggregated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_cor.html">run_gene_cor</a></span><span class="op">(</span></span>
<span>    <span class="va">data</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes_of_interest</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># we can change this number to pick 100 top variable genes</span></span>
<span><span class="va">feature_gene_prop_aggregated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_prop.html">run_gene_prop</a></span><span class="op">(</span><span class="va">data</span>, num_top_gene <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="classification-of-conditions-using-the-generated-features">Classification of conditions using the generated features<a class="anchor" aria-label="anchor" href="#classification-of-conditions-using-the-generated-features"></a>
</h3>
<p>To build disease prediction model from the generated features use
utilise <a href="https://www.bioconductor.org/packages/release/bioc/html/ClassifyR.html" class="external-link"><code>ClassifyR</code></a>.</p>
<p>The output from scFeatures is a matrix of sample x feature, ie, the
row corresponds to each sample, the column corresponds to the feature,
and can be directly used as the <code>X</code>. The order of the rows is
in the order of unique(data$sample).</p>
<p>Here we use the <code>condition</code> as the <code>y</code> to build
classification model on the disease condition.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using the feature type gene mean celltype as an example")</span></span>
<span></span>
<span><span class="co"># inspect the first 5 rows and first 5 columns</span></span>
<span><span class="va">feature_gene_mean_celltype</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                            Naive T Cells--NOSIP Naive T Cells--LUC7L3</span></span>
<span><span class="co">#&gt; Pre_P8_cond_Responder                 1.3403199             1.3253486</span></span>
<span><span class="co">#&gt; Pre_P27_cond_Non-responder            0.9504078             0.9565850</span></span>
<span><span class="co">#&gt; Pre_P7_cond_Responder                 0.9230278             0.7445286</span></span>
<span><span class="co">#&gt; Pre_P20_cond_Non-responder            0.9472109             0.9359819</span></span>
<span><span class="co">#&gt; Pre_P3_cond_Non-responder             0.7099657             1.3188926</span></span>
<span><span class="co">#&gt;                            Naive T Cells--ARL6IP4 Naive T Cells--IL2RG</span></span>
<span><span class="co">#&gt; Pre_P8_cond_Responder                   1.1861664             1.304897</span></span>
<span><span class="co">#&gt; Pre_P27_cond_Non-responder              0.0000000             1.018693</span></span>
<span><span class="co">#&gt; Pre_P7_cond_Responder                   0.2737271             1.462253</span></span>
<span><span class="co">#&gt; Pre_P20_cond_Non-responder              0.4789126             1.944960</span></span>
<span><span class="co">#&gt; Pre_P3_cond_Non-responder               0.6994015             1.607346</span></span>
<span><span class="co">#&gt;                            Naive T Cells--SNRPD2</span></span>
<span><span class="co">#&gt; Pre_P8_cond_Responder                  1.1722461</span></span>
<span><span class="co">#&gt; Pre_P27_cond_Non-responder             0.0000000</span></span>
<span><span class="co">#&gt; Pre_P7_cond_Responder                  0.8139368</span></span>
<span><span class="co">#&gt; Pre_P20_cond_Non-responder             0.4753097</span></span>
<span><span class="co">#&gt; Pre_P3_cond_Non-responder              0.4793805</span></span>
<span></span>
<span><span class="co"># inspect the dimension of the matrix</span></span>
<span><span class="co"># this means that there are 12354 features in this feature type</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">feature_gene_mean_celltype</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]   16 3642</span></span>
<span></span>
<span><span class="co"># confirm that the rows in in the order of unique(data$sample)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">feature_gene_mean_celltype</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>We recommend using <code><a href="https://rdrr.io/pkg/ClassifyR/man/crossValidate.html" class="external-link">ClassifyR::crossValidate</a></code> to do
cross-validated classification with the extracted feaures.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ClassifyR</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># X is the feature type generated</span></span>
<span><span class="co"># y is the condition for classification</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="va">feature_gene_mean_celltype</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span>, <span class="va">y</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span>, <span class="op">]</span><span class="op">$</span><span class="va">condition</span></span>
<span></span>
<span><span class="co"># run the classification model using random forest</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">ClassifyR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ClassifyR/man/crossValidate.html" class="external-link">crossValidate</a></span><span class="op">(</span></span>
<span>    <span class="va">X</span>, <span class="va">y</span>,</span>
<span>    classifier <span class="op">=</span> <span class="st">"randomForest"</span>, nCores <span class="op">=</span> <span class="fl">8</span>,</span>
<span>    nFolds <span class="op">=</span> <span class="fl">3</span>, nRepeats <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ClassifyR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ClassifyR/man/performancePlot.html" class="external-link">performancePlot</a></span><span class="op">(</span>results <span class="op">=</span> <span class="va">result</span><span class="op">)</span></span></code></pre></div>
<p><img src="scFeatures_detail_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
</div>
</div>
<div class="section level2">
<h2 id="scfeatures-on-spatial-proteomics-data">scFeatures on spatial proteomics data<a class="anchor" aria-label="anchor" href="#scfeatures-on-spatial-proteomics-data"></a>
</h2>
<p>For demonstration purpose, we “generate” a spatial proteomics data
from the single cell RNA-seq data by randomly generating some x and y -
coordinates.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> </span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_scrnaseq.rds"</span>, package <span class="op">=</span> <span class="st">"scFeatures"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># randomly generate some x and y coordinate to make this a </span></span>
<span><span class="co"># toy "spatial proteomics" data</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> , replace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> , replace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSeurat.html">makeSeurat</a></span><span class="op">(</span><span class="va">data</span>, spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">data</span>, normalise <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>  </span>
<span></span>
<span><span class="co"># how the x- and y- coordinates look like </span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">x_cord</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;   E6_P3_MMD4_L001    B2_P3_M91_L001 H1_P7_M60.B1_L001    E3_P1_M53_L001 </span></span>
<span><span class="co">#&gt;                37                 3                81                21 </span></span>
<span><span class="co">#&gt;         G3_P2_M17 </span></span>
<span><span class="co">#&gt;                38</span></span></code></pre></div>
<div class="section level3">
<h3 id="feature-generation-1">Feature generation<a class="anchor" aria-label="anchor" href="#feature-generation-1"></a>
</h3>
<div class="section level4">
<h4 id="cell-type-specific-proportions">Cell type specific proportions<a class="anchor" aria-label="anchor" href="#cell-type-specific-proportions"></a>
</h4>
<p>For spatial proteomics, we need to set the <code>type</code> to
<code>spatial_p</code></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_proportion_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_raw.html">run_proportion_raw</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_proportion_logit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_logit.html">run_proportion_logit</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_proportion_ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_ratio.html">run_proportion_ratio</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="cell-type-specific-gene-expressions-1">Cell type specific gene expressions<a class="anchor" aria-label="anchor" href="#cell-type-specific-gene-expressions-1"></a>
</h4>
<p>By default, this feature category calculates the gene/protein
expression based on the top 100 most variable genes/proteins per cell
type. However, because this data only measured 38 proteins, all proteins
will be used.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_gene_mean_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean_celltype.html">run_gene_mean_celltype</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_gene_prop_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_prop_celltype.html">run_gene_prop_celltype</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_gene_cor_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_cor_celltype.html">run_gene_cor_celltype</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="overall-aggregated-gene-expressions-1">Overall aggregated gene expressions<a class="anchor" aria-label="anchor" href="#overall-aggregated-gene-expressions-1"></a>
</h4>
<p>By default, this feature category calculates the gene/protein
expression based on the top 1500 most variable genes/proteins. However,
because in this data there are only 38 proteins, we will use all
proteins.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_gene_mean_aggregated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean.html">run_gene_mean</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_gene_prop_aggregated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_prop.html">run_gene_prop</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_gene_cor_aggregated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_cor.html">run_gene_cor</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="spatial-metrics">Spatial metrics<a class="anchor" aria-label="anchor" href="#spatial-metrics"></a>
</h4>
<ul>
<li>Detail:<br>
This feature category is designed for spatial data and contains feature
types commonly used for spatial analysis.
<ul>
<li>
<code>L_function</code>: the L values between the pairs of proteins
are calculated using the L function defined in literature[2]. L value
greater than zero indicates spatial attraction of the pair of proteins
whereas L value less than zero indicates spatial repulsion.</li>
<li>
<code>Morans_I</code>: Moran’s I are calculated using the function
defined in literature [3]. It calculates the spatial autocorrelation
based on both the locations and values simultaneously. A value closer to
1 indicates clustering of similar values and a value closer to -1
indicates clustering of dissimilar values. A value of 0 indicates no
particular clustering structure, ie, the values are spatially
distributed randomly.<br>
</li>
<li>
<code>celltype_interaction</code>: We find the nearest neighbours of
each cell and the cell types of these neighbours. These are considered
as spatial interaction pairs. The cell type composition of the spatial
interaction pairs are used as features.<br>
</li>
<li>
<code>nn_correlation</code>: Pearson’s correlation is calculated for
the protein expression between a cell with its nearest neighbour cell
for spatial proteomics.</li>
</ul>
</li>
<li>Expected output:
<ul>
<li>
<code>L_function</code>: The features are in the form of protein 1
vs protein 2, protein 1 vs protein 3 … etc, with the numbers
representing the L values.<br>
</li>
<li>
<code>Morans_I</code>: The features are in the form of protein 1,
protein 2 … etc, with the numbers representing Moran’s value.</li>
<li>
<code>celltype_interaction</code>: The features are in the form of
protein 1 vs protein 2, protein 1 vs protein 3 … etc, with the numbers
representing the proportion of each interaction pairs in a give
sample.</li>
<li>
<code>nn_correlation</code>: The features are in the form of protein
1, protein 2 … etc, with the numbers representing Pearson’s
correlation.</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_L_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_L_function.html">run_L_function</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_morans_I</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_Morans_I.html">run_Morans_I</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_celltype_interaction</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_celltype_interaction.html">run_celltype_interaction</a></span><span class="op">(</span></span>
<span>    <span class="va">data</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"spatial_p"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">feature_nn_correlation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_nn_correlation.html">run_nn_correlation</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="survival-analysis-using-the-generated-features">Survival analysis using the generated features<a class="anchor" aria-label="anchor" href="#survival-analysis-using-the-generated-features"></a>
</h3>
<p>Suppose we want to use the features to perform survival analysis,
here we “generate” some random survival data.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/therneau/survival" class="external-link">survival</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/survminer/index.html" class="external-link">survminer</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">nncor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">feature_nn_correlation</span><span class="op">)</span></span>
<span><span class="va">nncor</span> <span class="op">&lt;-</span> <span class="va">nncor</span><span class="op">[</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">nncor</span><span class="op">)</span><span class="op">==</span><span class="fl">0</span><span class="op">]</span></span>
<span><span class="co"># run hierarchical clustering</span></span>
<span><span class="va">hclust_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">as.dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">nncor</span>, method <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"ward.D2"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># generate some survival outcome</span></span>
<span><span class="va">survival_day</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">nncor</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">censoring</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">nncor</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cluster_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html" class="external-link">cutree</a></span><span class="op">(</span><span class="va">hclust_res</span>, k <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cluster_res</span><span class="op">)</span>,</span>
<span>                        survival_day <span class="op">=</span> <span class="va">survival_day</span>,</span>
<span>                        censoring <span class="op">=</span> <span class="va">censoring</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot survival curve</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/survfit.html" class="external-link">survfit</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">survival_day</span>, <span class="va">censoring</span><span class="op">)</span> <span class="op">~</span> <span class="va">cluster</span>,</span>
<span>    data <span class="op">=</span> <span class="va">metadata</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ggsurv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/survminer/man/ggsurvplot.html" class="external-link">ggsurvplot</a></span><span class="op">(</span><span class="va">fit</span>,</span>
<span>    conf.int <span class="op">=</span> <span class="cn">FALSE</span>, risk.table <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    risk.table.col <span class="op">=</span> <span class="st">"strata"</span>, pval <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ggsurv</span></span></code></pre></div>
<p><img src="scFeatures_detail_files/figure-html/unnamed-chunk-20-1.png" width="672"></p>
</div>
</div>
<div class="section level2">
<h2 id="scfeatures-on-spatial-transcriptomics-data">scFeatures on spatial transcriptomics data<a class="anchor" aria-label="anchor" href="#scfeatures-on-spatial-transcriptomics-data"></a>
</h2>
<p>For demonstration purpose, we also “generate” the spatial
transciptomics from the single-cell RNA-seq data.</p>
<p>For spatial transcriptomics, the required information are
<code>sample</code>, <code>x_cord</code> and <code>y_cord</code> (the x-
and y- coordinates of each spot).</p>
<p>Additionally, there are two assays needed, the <code>RNA</code>,
which contains the gene expression of each spot and the
<code>predictions</code>.</p>
<p>The <code>predictions</code> assay is a matrix in the form of cell
types x spot, which stores the cell type probability of each spot. This
can be obtained from performing cell type prediction using reference
data, for example, using <code>SCTransform</code> from Seurat (see <a href="https://satijalab.org/seurat/articles/spatial_vignette.html#integration-with-single-cell-data-1" class="external-link uri">https://satijalab.org/seurat/articles/spatial_vignette.html#integration-with-single-cell-data-1</a>).
The purpose of having this cell type probability matrix is due to
spatial transcriptomics assay contain multiple cells from multiple cell
type populations, and this cell type prediction is necessary for
generating cell type specific feature types.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_scrnaseq.rds"</span>, package <span class="op">=</span> <span class="st">"scFeatures"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># generate a toy "spatial transcriptomics" data</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="cn">NULL</span> <span class="co">#spatial transcriptomics don't have celltype </span></span>
<span></span>
<span><span class="co"># randomly generate some x and y coordinates</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> , replace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> , replace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># for spatial transcriptomics, we need to estimate the number of cells per spot</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_num_cell_per_spot.html">get_num_cell_per_spot</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># also need a dataframe of celltype probability at each spot </span></span>
<span><span class="co"># here we randomly create one </span></span>
<span><span class="va">nrow</span> <span class="op">&lt;-</span> <span class="fl">5</span> <span class="co">#pretend there are 5 cell types </span></span>
<span><span class="va">ncol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Create a matrix of random numbers</span></span>
<span><span class="va">matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="va">nrow</span> <span class="op">*</span> <span class="va">ncol</span><span class="op">)</span>, <span class="va">nrow</span>, <span class="va">ncol</span><span class="op">)</span></span>
<span><span class="co"># Normalize the columns of the matrix so that each column sum to 1</span></span>
<span><span class="va">prediction.scores</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">matrix</span>, <span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">matrix</span><span class="op">)</span>, <span class="st">"/"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">prediction.scores</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">prediction.scores</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"celltype"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># format the data using makeSeurat</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSeurat.html">makeSeurat</a></span><span class="op">(</span><span class="va">data</span>, spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span><span class="op">)</span>, spotProbability <span class="op">=</span> <span class="va">prediction.scores</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">data</span>, normalise <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>  </span>
<span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co"># how the prediction assay look like</span></span>
<span><span class="co"># the rows contain the predicted cell type probability of each spot</span></span>
<span><span class="va">data</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">predictions</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="co">#&gt;                            E6_P3_MMD4_L001</span></span>
<span><span class="co">#&gt; prediction.score.celltype1      0.06454154</span></span>
<span><span class="co">#&gt; prediction.score.celltype2      0.18199502</span></span>
<span><span class="co">#&gt; prediction.score.celltype3      0.19232122</span></span>
<span><span class="co">#&gt; prediction.score.celltype4      0.38186122</span></span>
<span><span class="co">#&gt; prediction.score.celltype5      0.17928100</span></span>
<span></span>
<span><span class="co"># confirm the data include samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "Pre_P8"  "Pre_P6"  "Pre_P27" "Pre_P7"  "Pre_P20"</span></span>
<span></span>
<span><span class="co"># conditions in this data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Responder"     "Non-responder"</span></span>
<span></span>
<span><span class="co"># format the patient ID as `a_cond_b` eg, `patient09_cond_nonresponder`. This is</span></span>
<span><span class="co"># so that we can retrieve the sample ID and the corresponding condition when</span></span>
<span><span class="co"># running downstream analysis on the generated features</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">sample</span>, <span class="st">"_cond_"</span>, <span class="va">data</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="feature-generation-2">Feature generation<a class="anchor" aria-label="anchor" href="#feature-generation-2"></a>
</h3>
<p>For spatial transcriptomics, each spot contains multiple cells and
potentially from multiple cell types. Therefore, the implementation for
the feature categories involving cell type specific features (i.e., cell
type proportions, cell type specific gene expressions, cell type
specific pathway expressions, spatial metrics) are different to their
implementations for the single-cell based datasets (i.e., scRNA-seq and
spatial proteomics) and described in details below.</p>
<div class="section level4">
<h4 id="cell-type-proportions-1">Cell type proportions<a class="anchor" aria-label="anchor" href="#cell-type-proportions-1"></a>
</h4>
<ul>
<li><p>Detail: While in spatial transriptomics, the measurement at each
spot is taken from multiple cells, we can use the cell type probability
matrix and the relative number of cells in each spot jointly to estimate
the cell type proportion in each sample.</p></li>
<li><p>Expected output:<br>
The output format is the same as for scRNA-seq data.</p></li>
</ul>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_proportion_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_raw.html">run_proportion_raw</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span><span class="va">feature_proportion_logit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_logit.html">run_proportion_logit</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span><span class="va">feature_proportion_ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_ratio.html">run_proportion_ratio</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="cell-type-specific-gene-expressions-2">Cell type specific gene expressions<a class="anchor" aria-label="anchor" href="#cell-type-specific-gene-expressions-2"></a>
</h4>
<ul>
<li><p>Detail: We use the cell type probability to obtain the regression
coefficients of each gene associated with each cell type.</p></li>
<li><p>Expected output: For each sample, the features are in the form of
gene a celltype a, gene a celltype b … etc, with the number representing
regression coefficient.</p></li>
</ul>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_remove_mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_mito.html">remove_mito</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="va">feature_gene_mean_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean_celltype.html">run_gene_mean_celltype</a></span><span class="op">(</span></span>
<span>    <span class="va">data_remove_mito</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"spatial_t"</span>,</span>
<span>    genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data_remove_mito</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span> <span class="co"># to speed up the vignette generation </span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="cell-type-specific-pathway-expressions-1">Cell type specific pathway expressions<a class="anchor" aria-label="anchor" href="#cell-type-specific-pathway-expressions-1"></a>
</h4>
<ul>
<li><p>Detail: We use the approach above to obtain the regression
coefficient of each cell type for each gene. The regression coefficients
for all the genes involved in a particular pathway are then
summed.</p></li>
<li><p>Expected output: For each sample, the features are in the form of
pathway a celltype a, pathway a celltype b … etc, with the number
representing the summation of regression coefficient.</p></li>
</ul>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_pathway_mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathway_mean.html">run_pathway_mean</a></span><span class="op">(</span></span>
<span>    <span class="va">data</span>,</span>
<span>    geneset <span class="op">=</span> <span class="cn">NULL</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="overall-aggregated-gene-expressions-2">Overall aggregated gene expressions<a class="anchor" aria-label="anchor" href="#overall-aggregated-gene-expressions-2"></a>
</h4>
<ul>
<li><p>Detail:<br>
This is based on the expression of genes across all cells.</p></li>
<li><p>Expected output:<br>
The output format is the same as for scRNA-seq data.</p></li>
</ul>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_gene_mean_aggregated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean.html">run_gene_mean</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span><span class="va">feature_gene_prop_aggregated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_prop.html">run_gene_prop</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span><span class="va">feature_gene_cor_aggregated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_cor.html">run_gene_cor</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="spatial-metrics-1">Spatial metrics<a class="anchor" aria-label="anchor" href="#spatial-metrics-1"></a>
</h4>
<ul>
<li>Detail:<br>
This feature category is designed for spatial data and contains feature
types commonly used for spatial analysis.
<ul>
<li>
<code>L_function</code>: the L values between the pairs of genes are
calculated using the L function defined in literature. We used the
estimated cell type proportion in each spot to calculate the L
function.<br>
</li>
<li>
<code>Morans_I</code>: The implementation is the same as for spatial
proteomics.</li>
<li>
<code>celltype_interaction</code>: We assume that the nearest
neighbours should be the cells captured within each spot and consider
them as the spatial interaction pairs. We then use the estimated cell
type proportion in each spot to calculate the spatial interaction
between cell types.<br>
</li>
<li>
<code>nn_correlation</code>: Pearson’s correlation is calculated for
the gene expression between a spot with its nearest neighbour spot.</li>
</ul>
</li>
<li>Expected output:
<ul>
<li>
<code>L_function</code>: The features are in the form of gene 1 vs
gene 2, gene 1 vs gene 3 … etc, with the numbers representing the L
values.<br>
</li>
<li>
<code>Morans_I</code>: The features are in the form of gene 1, gene
2 … etc, with the numbers representing Moran’s value.</li>
<li>
<code>celltype_interaction</code>: The features are in the form of
gene 1 vs gene 2, gene 1 vs gene 3 … etc, with the numbers representing
the proportion of each interaction pairs in a give sample.</li>
<li>
<code>nn_correlation</code>: The features are in the form of gene 1,
gene 2 … etc, with the numbers representing Pearson’s correlation.</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_L_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_L_function.html">run_L_function</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span><span class="va">feature_morans_I</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_Morans_I.html">run_Morans_I</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span><span class="va">feature_celltype_interaction</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_celltype_interaction.html">run_celltype_interaction</a></span><span class="op">(</span></span>
<span>    <span class="va">data</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"spatial_t"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">feature_nn_correlation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_nn_correlation.html">run_nn_correlation</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Sade-Feldman, M., Yizhak, K., Bjorgaard, S. L., Ray, J. P., de Boer,
C. G., Jenkins, R. W., … &amp; Hacohen, N.(2018). Defining T cell states
associated with response to checkpoint immunotherapy in melanoma. Cell,
175(4), 998-1013.</li>
<li>Besag, J. (1977) Discussion of Dr Ripley’s paper. Journal of the
Royal Statistical Society, Series B, 39, 193–195.</li>
<li>Moran, P. A. P. (1950).”Notes on Continuous Stochastic Phenomena”.
Biometrika.37 (1) : 17–23. <a href="doi:10.2307/2332142" class="uri">doi:10.2307/2332142</a>
</li>
</ol>
</div>
<div class="section level2">
<h2 id="sessioninfo">sessionInfo()<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Debian GNU/Linux 11 (bullseye)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.13.so</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] survminer_0.4.9             ggpubr_0.4.0               </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.3.6               ClassifyR_3.1.28           </span></span>
<span><span class="co">#&gt;  [5] survival_3.4-0              BiocParallel_1.30.4        </span></span>
<span><span class="co">#&gt;  [7] MultiAssayExperiment_1.22.0 SummarizedExperiment_1.26.1</span></span>
<span><span class="co">#&gt;  [9] Biobase_2.56.0              GenomicRanges_1.48.0       </span></span>
<span><span class="co">#&gt; [11] GenomeInfoDb_1.32.4         IRanges_2.30.1             </span></span>
<span><span class="co">#&gt; [13] MatrixGenerics_1.8.1        matrixStats_0.62.0         </span></span>
<span><span class="co">#&gt; [15] generics_0.1.3              scFeatures_0.99.9          </span></span>
<span><span class="co">#&gt; [17] S4Vectors_0.34.0            BiocGenerics_0.42.0        </span></span>
<span><span class="co">#&gt; [19] BiocStyle_2.24.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] rsvd_1.0.5                  ica_1.0-3                  </span></span>
<span><span class="co">#&gt;   [3] Rsamtools_2.12.0            foreach_1.5.2              </span></span>
<span><span class="co">#&gt;   [5] SingleCellSignalR_1.8.0     lmtest_0.9-40              </span></span>
<span><span class="co">#&gt;   [7] rprojroot_2.0.3             crayon_1.5.2               </span></span>
<span><span class="co">#&gt;   [9] spatstat.core_2.4-4         MASS_7.3-58.1              </span></span>
<span><span class="co">#&gt;  [11] rhdf5filters_1.8.0          nlme_3.1-160               </span></span>
<span><span class="co">#&gt;  [13] backports_1.4.1             rlang_1.0.6                </span></span>
<span><span class="co">#&gt;  [15] XVector_0.36.0              ROCR_1.0-11                </span></span>
<span><span class="co">#&gt;  [17] irlba_2.3.5.1               limma_3.52.4               </span></span>
<span><span class="co">#&gt;  [19] filelock_1.0.2              rjson_0.2.21               </span></span>
<span><span class="co">#&gt;  [21] bit64_4.0.5                 glue_1.6.2                 </span></span>
<span><span class="co">#&gt;  [23] pheatmap_1.0.12             sctransform_0.3.5          </span></span>
<span><span class="co">#&gt;  [25] parallel_4.2.1              spatstat.sparse_3.0-0      </span></span>
<span><span class="co">#&gt;  [27] AnnotationDbi_1.58.0        spatstat.geom_3.0-3        </span></span>
<span><span class="co">#&gt;  [29] tidyselect_1.2.0            km.ci_0.5-6                </span></span>
<span><span class="co">#&gt;  [31] SeuratObject_4.1.2          fitdistrplus_1.1-8         </span></span>
<span><span class="co">#&gt;  [33] XML_3.99-0.11               tidyr_1.2.1                </span></span>
<span><span class="co">#&gt;  [35] zoo_1.8-11                  GenomicAlignments_1.32.1   </span></span>
<span><span class="co">#&gt;  [37] xtable_1.8-4                magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;  [39] evaluate_0.17               scuttle_1.6.3              </span></span>
<span><span class="co">#&gt;  [41] cli_3.4.1                   zlibbioc_1.42.0            </span></span>
<span><span class="co">#&gt;  [43] rstudioapi_0.14             miniUI_0.1.1.1             </span></span>
<span><span class="co">#&gt;  [45] sp_1.5-0                    bslib_0.4.0                </span></span>
<span><span class="co">#&gt;  [47] rpart_4.1.16                ensembldb_2.20.2           </span></span>
<span><span class="co">#&gt;  [49] shiny_1.7.2                 GSVA_1.44.5                </span></span>
<span><span class="co">#&gt;  [51] BiocSingular_1.12.0         xfun_0.33                  </span></span>
<span><span class="co">#&gt;  [53] multtest_2.52.0             cluster_2.1.4              </span></span>
<span><span class="co">#&gt;  [55] caTools_1.18.2              ggtext_0.1.2               </span></span>
<span><span class="co">#&gt;  [57] KEGGREST_1.36.3             tibble_3.1.8               </span></span>
<span><span class="co">#&gt;  [59] ggrepel_0.9.1               ape_5.6-2                  </span></span>
<span><span class="co">#&gt;  [61] listenv_0.8.0               Biostrings_2.64.1          </span></span>
<span><span class="co">#&gt;  [63] png_0.1-7                   future_1.28.0              </span></span>
<span><span class="co">#&gt;  [65] withr_2.5.0                 bitops_1.0-7               </span></span>
<span><span class="co">#&gt;  [67] ranger_0.14.1               plyr_1.8.7                 </span></span>
<span><span class="co">#&gt;  [69] GSEABase_1.58.0             AnnotationFilter_1.20.0    </span></span>
<span><span class="co">#&gt;  [71] pracma_2.4.2                dqrng_0.3.0                </span></span>
<span><span class="co">#&gt;  [73] pillar_1.8.1                RcppParallel_5.1.5         </span></span>
<span><span class="co">#&gt;  [75] gplots_3.1.3                GlobalOptions_0.1.2        </span></span>
<span><span class="co">#&gt;  [77] cachem_1.0.6                GenomicFeatures_1.48.4     </span></span>
<span><span class="co">#&gt;  [79] fs_1.5.2                    DelayedMatrixStats_1.18.1  </span></span>
<span><span class="co">#&gt;  [81] vctrs_0.4.2                 ellipsis_0.3.2             </span></span>
<span><span class="co">#&gt;  [83] tools_4.2.1                 munsell_0.5.0              </span></span>
<span><span class="co">#&gt;  [85] DelayedArray_0.22.0         fastmap_1.1.0              </span></span>
<span><span class="co">#&gt;  [87] compiler_4.2.1              abind_1.4-5                </span></span>
<span><span class="co">#&gt;  [89] httpuv_1.6.6                rtracklayer_1.56.1         </span></span>
<span><span class="co">#&gt;  [91] plotly_4.10.0               rgeos_0.5-9                </span></span>
<span><span class="co">#&gt;  [93] GenomeInfoDbData_1.2.8      gridExtra_2.3              </span></span>
<span><span class="co">#&gt;  [95] edgeR_3.38.4                lattice_0.20-45            </span></span>
<span><span class="co">#&gt;  [97] deldir_1.0-6                utf8_1.2.2                 </span></span>
<span><span class="co">#&gt;  [99] later_1.3.0                 dplyr_1.0.10               </span></span>
<span><span class="co">#&gt; [101] BiocFileCache_2.4.0         jsonlite_1.8.2             </span></span>
<span><span class="co">#&gt; [103] scales_1.2.1                graph_1.74.0               </span></span>
<span><span class="co">#&gt; [105] ScaledMatrix_1.4.1          carData_3.0-5              </span></span>
<span><span class="co">#&gt; [107] pbapply_1.5-0               sparseMatrixStats_1.8.0    </span></span>
<span><span class="co">#&gt; [109] genefilter_1.78.0           lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt; [111] promises_1.2.0.1            car_3.1-0                  </span></span>
<span><span class="co">#&gt; [113] R.utils_2.12.0              goftest_1.2-3              </span></span>
<span><span class="co">#&gt; [115] spatstat.utils_3.0-1        reticulate_1.26            </span></span>
<span><span class="co">#&gt; [117] rmarkdown_2.17              pkgdown_2.0.6              </span></span>
<span><span class="co">#&gt; [119] cowplot_1.1.1               textshaping_0.3.6          </span></span>
<span><span class="co">#&gt; [121] statmod_1.4.37              Rtsne_0.16                 </span></span>
<span><span class="co">#&gt; [123] uwot_0.1.14                 igraph_1.3.5               </span></span>
<span><span class="co">#&gt; [125] HDF5Array_1.24.2            proxyC_0.3.3               </span></span>
<span><span class="co">#&gt; [127] yaml_2.3.5                  systemfonts_1.0.4          </span></span>
<span><span class="co">#&gt; [129] htmltools_0.5.3             memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [131] BiocIO_1.6.0                Seurat_4.2.0               </span></span>
<span><span class="co">#&gt; [133] locfit_1.5-9.6              viridisLite_0.4.1          </span></span>
<span><span class="co">#&gt; [135] digest_0.6.29               assertthat_0.2.1           </span></span>
<span><span class="co">#&gt; [137] mime_0.12                   rappdirs_0.3.3             </span></span>
<span><span class="co">#&gt; [139] KMsurv_0.1-5                SIMLR_1.22.0               </span></span>
<span><span class="co">#&gt; [141] RSQLite_2.2.18              future.apply_1.9.1         </span></span>
<span><span class="co">#&gt; [143] data.table_1.14.2           blob_1.2.3                 </span></span>
<span><span class="co">#&gt; [145] R.oo_1.25.0                 survMisc_0.5.6             </span></span>
<span><span class="co">#&gt; [147] ragg_1.2.3                  splines_4.2.1              </span></span>
<span><span class="co">#&gt; [149] labeling_0.4.2              Rhdf5lib_1.18.2            </span></span>
<span><span class="co">#&gt; [151] ProtGenerics_1.28.0         gridtext_0.1.5             </span></span>
<span><span class="co">#&gt; [153] RCurl_1.98-1.9              broom_1.0.1                </span></span>
<span><span class="co">#&gt; [155] hms_1.1.2                   rhdf5_2.40.0               </span></span>
<span><span class="co">#&gt; [157] colorspace_2.0-3            DropletUtils_1.16.0        </span></span>
<span><span class="co">#&gt; [159] BiocManager_1.30.18         shape_1.4.6                </span></span>
<span><span class="co">#&gt; [161] sass_0.4.2                  Rcpp_1.0.9                 </span></span>
<span><span class="co">#&gt; [163] bookdown_0.30               RANN_2.6.1                 </span></span>
<span><span class="co">#&gt; [165] circlize_0.4.15             fansi_1.0.3                </span></span>
<span><span class="co">#&gt; [167] parallelly_1.32.1           R6_2.5.1                   </span></span>
<span><span class="co">#&gt; [169] grid_4.2.1                  ggridges_0.5.4             </span></span>
<span><span class="co">#&gt; [171] lifecycle_1.0.3             bluster_1.6.0              </span></span>
<span><span class="co">#&gt; [173] curl_4.3.3                  ggsignif_0.6.3             </span></span>
<span><span class="co">#&gt; [175] leiden_0.4.3                jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt; [177] SpatialExperiment_1.6.1     Matrix_1.5-3               </span></span>
<span><span class="co">#&gt; [179] EnsDb.Mmusculus.v79_2.99.0  desc_1.4.2                 </span></span>
<span><span class="co">#&gt; [181] RcppAnnoy_0.0.19            RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt; [183] iterators_1.0.14            spatstat.explore_3.0-5     </span></span>
<span><span class="co">#&gt; [185] stringr_1.4.1               htmlwidgets_1.5.4          </span></span>
<span><span class="co">#&gt; [187] markdown_1.1                beachmat_2.12.0            </span></span>
<span><span class="co">#&gt; [189] polyclip_1.10-0             biomaRt_2.52.0             </span></span>
<span><span class="co">#&gt; [191] purrr_0.3.5                 mgcv_1.8-40                </span></span>
<span><span class="co">#&gt; [193] globals_0.16.1              patchwork_1.1.2            </span></span>
<span><span class="co">#&gt; [195] spatstat.random_3.0-1       progressr_0.11.0           </span></span>
<span><span class="co">#&gt; [197] codetools_0.2-18            metapod_1.4.0              </span></span>
<span><span class="co">#&gt; [199] gtools_3.9.3                prettyunits_1.1.1          </span></span>
<span><span class="co">#&gt; [201] SingleCellExperiment_1.18.1 dbplyr_2.2.1               </span></span>
<span><span class="co">#&gt; [203] EnsDb.Hsapiens.v79_2.99.0   RSpectra_0.16-1            </span></span>
<span><span class="co">#&gt; [205] R.methodsS3_1.8.2           gtable_0.3.1               </span></span>
<span><span class="co">#&gt; [207] DBI_1.1.3                   tensor_1.5                 </span></span>
<span><span class="co">#&gt; [209] httr_1.4.4                  highr_0.9                  </span></span>
<span><span class="co">#&gt; [211] KernSmooth_2.23-20          stringi_1.7.8              </span></span>
<span><span class="co">#&gt; [213] progress_1.2.2              reshape2_1.4.4             </span></span>
<span><span class="co">#&gt; [215] msigdbr_7.5.1               farver_2.1.1               </span></span>
<span><span class="co">#&gt; [217] annotate_1.74.0             magick_2.7.3               </span></span>
<span><span class="co">#&gt; [219] xml2_1.3.3                  BiocNeighbors_1.14.0       </span></span>
<span><span class="co">#&gt; [221] AUCell_1.18.1               restfulr_0.0.15            </span></span>
<span><span class="co">#&gt; [223] scattermore_0.8             scran_1.24.1               </span></span>
<span><span class="co">#&gt; [225] bit_4.0.4                   spatstat.data_3.0-0        </span></span>
<span><span class="co">#&gt; [227] pkgconfig_2.0.3             babelgene_22.9             </span></span>
<span><span class="co">#&gt; [229] rstatix_0.7.0               knitr_1.40</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yue Cao, Yingxin Lin, Ellis Patrick, Pengyi Yang, Jean Yee Hwa Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
