<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scFeatures">
<title>An overview of scFeatures' functions â€¢ scFeatures</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="An overview of scFeatures' functions">
<meta property="og:description" content="scFeatures">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scFeatures</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.9</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/scFeatures_associationstudy.html">An association study of features with conditions with scFeatures</a>
    <a class="dropdown-item" href="../articles/scFeatures_detail.html">A detailed explanation of scFeatures' features</a>
    <a class="dropdown-item" href="../articles/scFeatures_summary.html">An overview of scFeatures' functions</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/SydneyBioX/scFeatures/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>An overview of scFeatures' functions</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/SydneyBioX/scFeatures/blob/HEAD/vignettes/scFeatures_summary.Rmd" class="external-link"><code>vignettes/scFeatures_summary.Rmd</code></a></small>
      <div class="d-none name"><code>scFeatures_summary.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette provides an overview of scFeatures package. scFeatures
generates features for single-cell RNA-seq, spatial proteomic and
spatial transcriptomic data. The features span across six categories
representing different molecular views of cellular characteristics.
These include (i) cell type proportions, (ii) cell type specific gene
expressions, (iii) cell type specific pathway expressions, (iv) cell
type specific cell-cell interaction (CCI) scores, (v) overall aggregated
gene expressions and (vi) spatial metrics. These features enable a more
comprehensive multi-view representation of the expression data and can
be used for downstream analysis such as association study.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scFeatures</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>scFeatures can be run using one line of code
<code>scfeatures_result &lt;- scFeatures(data)</code>, which generates a
list of data frames containing all feature types in the form of samples
x features.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_scrnaseq.rds"</span>, package <span class="op">=</span> <span class="st">"scFeatures"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">data</span>, normalise <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co">#  perform normalisation</span></span>
<span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scFeatures.html">scFeatures</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<p>Below we list the function for generating each individual feature
types for scRNA-seq, spatial proteomic and spatial transcriptomic
data.</p>
</div>
<div class="section level2">
<h2 id="scfeatures-for-single-cell-rna-seq-data">scFeatures for single-cell RNA-seq data<a class="anchor" aria-label="anchor" href="#scfeatures-for-single-cell-rna-seq-data"></a>
</h2>
<p>scFeatures accept a Seurat object containing the <code>sample</code>
and <code>celltype</code> columns. For SingleCellExperiment or
SpatialExperiment object, we provide a function
<code><a href="../reference/makeSeurat.html">makeSeurat()</a></code> for the conversion.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># read in data</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_scrnaseq.rds"</span>, package <span class="op">=</span> <span class="st">"scFeatures"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">data</span>, normalise <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co">#  perform normalisation</span></span>
<span></span>
<span><span class="co"># format the patient ID as `a_cond_b` eg, `patient09_cond_nonresponder`.</span></span>
<span><span class="co"># This is so that we can retrieve the sample ID and the corresponding condition</span></span>
<span><span class="co"># when running downstream analysis on the generated features</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">sample</span>, <span class="st">"_cond_"</span>, <span class="va">data</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type proportions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_proportion_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_raw.html">run_proportion_raw</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">feature_proportion_logit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_logit.html">run_proportion_logit</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">feature_proportion_ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_ratio.html">run_proportion_ratio</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type specific gene expressions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># optional step, run if mitochondria amd ribosomal genes are not of interest</span></span>
<span><span class="va">data_remove_mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_mito.html">remove_mito</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># by default, use the top variable genes</span></span>
<span><span class="co"># users can change the number of top variable genes by changing `num_top_gene`</span></span>
<span><span class="co"># alternatively, users can also input genes of interest in a dataframe format</span></span>
<span><span class="va">genes_of_interest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    marker <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"S100A11"</span>, <span class="st">"GZMB"</span>, <span class="st">"DUSP1"</span><span class="op">)</span>,</span>
<span>    celltype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"CD8, T Effector Memory"</span>, </span>
<span>      <span class="st">"CD8, T Effector Memory"</span>, </span>
<span>      <span class="st">"Naive T Cells"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">feature_gene_mean_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean_celltype.html">run_gene_mean_celltype</a></span><span class="op">(</span><span class="va">data_remove_mito</span><span class="op">)</span></span>
<span><span class="va">feature_gene_prop_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_prop_celltype.html">run_gene_prop_celltype</a></span><span class="op">(</span></span>
<span>  <span class="va">data_remove_mito</span>, genes <span class="op">=</span> <span class="va">genes_of_interest</span></span>
<span><span class="op">)</span></span>
<span><span class="va">feature_gene_cor_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_cor_celltype.html">run_gene_cor_celltype</a></span><span class="op">(</span></span>
<span>  <span class="va">data_remove_mito</span>, num_top_gene <span class="op">=</span> <span class="fl">50</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># by default, pick around 100 most variable genes per cell type, can change</span></span>
<span><span class="co"># this number through the num_top_gene argument</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type specific pathway expressions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_pathway_gsva</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathway_gsva.html">run_pathway_gsva</a></span><span class="op">(</span><span class="va">data</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span><span class="op">)</span></span>
<span><span class="va">feature_pathway_mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathway_mean.html">run_pathway_mean</a></span><span class="op">(</span><span class="va">data</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span><span class="op">)</span></span>
<span><span class="co"># by default, use the 50 hallmark pathways, users can also input their gene</span></span>
<span><span class="co"># set of interest in a list format</span></span>
<span><span class="va">geneset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"pathway_a"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CAPNS1"</span>, <span class="st">"TLCD1"</span><span class="op">)</span>,</span>
<span>    <span class="st">"pathway_b"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PEX6"</span>, <span class="st">"DPRXP4"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">feature_pathway_prop</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathway_prop.html">run_pathway_prop</a></span><span class="op">(</span></span>
<span>  <span class="va">data</span>, geneset <span class="op">=</span> <span class="va">geneset</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"Homo sapiens"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type specific cell-cell interactions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_CCI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_CCI.html">run_CCI</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># Bulk expressions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span></span>
<span><span class="co"># by default, use the top variable genes</span></span>
<span><span class="co"># users can change the number of top variable genes by changing `num_top_gene`</span></span>
<span><span class="co"># alternatively, users can also input their genes of interest in a list format</span></span>
<span><span class="va">feature_gene_mean_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean.html">run_gene_mean</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">feature_gene_cor_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_cor.html">run_gene_cor</a></span><span class="op">(</span><span class="va">data</span>, num_top_gene <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">genes_of_interest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TIGIT"</span>, <span class="st">"PDCD1"</span><span class="op">)</span></span>
<span><span class="va">feature_gene_prop_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_prop.html">run_gene_prop</a></span><span class="op">(</span><span class="va">data</span>, genes <span class="op">=</span> <span class="va">genes_of_interest</span><span class="op">)</span></span>
<span><span class="co"># by default, pick 1500 most variable genes, can change this number through the</span></span>
<span><span class="co"># num_top_gene argument</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="scfeatures-for-spatial-proteomics">scFeatures for spatial proteomics<a class="anchor" aria-label="anchor" href="#scfeatures-for-spatial-proteomics"></a>
</h2>
<p>The input data is a Seurat object containing <code>celltype</code>,
<code>sample</code>, <code>x_coord</code> and <code>y_coord</code>
columns.</p>
<p>For SingleCellExperiment or SpatialExperiment object, we provide a
function <code><a href="../reference/makeSeurat.html">makeSeurat()</a></code> for the conversion. use</p>
<p>Note that, as spatial proteomics contain few genes, the feature
categories of pathway expressions and cell-cell interactions are not
applicable.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># read in data</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_scrnaseq.rds"</span>, package <span class="op">=</span> <span class="st">"scFeatures"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># randomly generate some x and y coordinate to make this a </span></span>
<span><span class="co"># toy "spatial proteomics" data</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> , replace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> , replace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSeurat.html">makeSeurat</a></span><span class="op">(</span><span class="va">data</span>, spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">data</span>, normalise <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>  </span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type proportions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_proportion_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_raw.html">run_proportion_raw</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_proportion_logit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_logit.html">run_proportion_logit</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_proportion_ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_ratio.html">run_proportion_ratio</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type specific gene expressions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_gene_mean_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean_celltype.html">run_gene_mean_celltype</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_gene_prop_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_prop_celltype.html">run_gene_prop_celltype</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_gene_cor_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_cor_celltype.html">run_gene_cor_celltype</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># Bulk expressions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_gene_mean_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean.html">run_gene_mean</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_gene_prop_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_prop.html">run_gene_prop</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_gene_cor_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_cor.html">run_gene_cor</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># Spatial metrics</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_L_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_L_function.html">run_L_function</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_morans_I</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_Morans_I.html">run_Morans_I</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_celltype_interact</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_celltype_interaction.html">run_celltype_interaction</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span>
<span><span class="va">feature_nn_correlation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_nn_correlation.html">run_nn_correlation</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_p"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="scfeature-for-spatial-transcriptomics">scFeature for spatial transcriptomics<a class="anchor" aria-label="anchor" href="#scfeature-for-spatial-transcriptomics"></a>
</h2>
<p>The input data is a Seurat object containing <code>sample</code>,
<code>x_cord</code> and <code>y_cord</code> columns.<br>
Additionally, apart from the <code>RNA</code> assay, which contains the
gene expression of each spot, a <code>predictions</code> assay is also
needed.</p>
<p>The <code>predictions</code> assay is a matrix in the form of cell
types x spot, which stores the cell type probability of each spot. This
can be obtained from performing cell type prediction using reference
data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># read in data</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_scrnaseq.rds"</span>, package <span class="op">=</span> <span class="st">"scFeatures"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># generate a toy "spatial transcriptomics" data</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="cn">NULL</span> <span class="co">#spatial transcriptomics don't have celltype </span></span>
<span></span>
<span><span class="co"># randomly generate some x and y coordinates</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> , replace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> , replace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># for spatial transcriptomics, we need to estimate the number of cells per spot</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_num_cell_per_spot.html">get_num_cell_per_spot</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># also need a dataframe of celltype probability at each spot </span></span>
<span><span class="co"># here we randomly create one </span></span>
<span><span class="va">nrow</span> <span class="op">&lt;-</span> <span class="fl">5</span> <span class="co">#pretend there are 5 cell types </span></span>
<span><span class="va">ncol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Create a matrix of random numbers</span></span>
<span><span class="va">matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="va">nrow</span> <span class="op">*</span> <span class="va">ncol</span><span class="op">)</span>, <span class="va">nrow</span>, <span class="va">ncol</span><span class="op">)</span></span>
<span><span class="co"># Normalize the columns of the matrix so that each column sum to 1</span></span>
<span><span class="va">prediction.scores</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">matrix</span>, <span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">matrix</span><span class="op">)</span>, <span class="st">"/"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">prediction.scores</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">prediction.scores</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"celltype"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># format the data using makeSeurat</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSeurat.html">makeSeurat</a></span><span class="op">(</span><span class="va">data</span>, spatialCoords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span><span class="op">)</span>, spotProbability <span class="op">=</span> <span class="va">prediction.scores</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">data</span>, normalise <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>  </span>
<span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co"># format the patient ID as `a_cond_b` eg, `patient09_cond_nonresponder`.</span></span>
<span><span class="co"># This is so that we can retrieve the sample ID and the corresponding</span></span>
<span><span class="co"># condition when running downstream analysis on the generated features</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">sample</span>, <span class="st">"_cond_"</span>, <span class="va">data</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type proportions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_proportion_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_raw.html">run_proportion_raw</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span><span class="va">feature_proportion_logit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_logit.html">run_proportion_logit</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span><span class="va">feature_proportion_ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_proportion_ratio.html">run_proportion_ratio</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type specific gene expressions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">data_remove_mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_mito.html">remove_mito</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">feature_gene_mean_celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean_celltype.html">run_gene_mean_celltype</a></span><span class="op">(</span></span>
<span>  <span class="va">data_remove_mito</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span>, genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data_remove_mito</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span> <span class="co">#to speed up example</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># cell type specific pathway expressions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_pathway_mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathway_mean.html">run_pathway_mean</a></span><span class="op">(</span><span class="va">data</span>, geneset <span class="op">=</span> <span class="cn">NULL</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span> </span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># Bulk expressions</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_gene_mean_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_mean.html">run_gene_mean</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span><span class="va">feature_gene_prop_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_prop.html">run_gene_prop</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span><span class="va">feature_gene_cor_bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_gene_cor.html">run_gene_cor</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="co"># Spatial metrics</span></span>
<span><span class="co">#---------------------------------------------------------------------------</span></span>
<span><span class="va">feature_L_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_L_function.html">run_L_function</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span><span class="va">feature_morans_I</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_Morans_I.html">run_Morans_I</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span><span class="va">feature_celltype_interact</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_celltype_interaction.html">run_celltype_interaction</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span>
<span><span class="va">feature_nn_correlation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_nn_correlation.html">run_nn_correlation</a></span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"spatial_t"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="sessioninfo">sessionInfo()<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Debian GNU/Linux 11 (bullseye)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.13.so</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] scFeatures_0.99.9 BiocStyle_2.24.0 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] rappdirs_0.3.3              rtracklayer_1.56.1         </span></span>
<span><span class="co">#&gt;   [3] scattermore_0.8             SpatialExperiment_1.6.1    </span></span>
<span><span class="co">#&gt;   [5] R.methodsS3_1.8.2           SeuratObject_4.1.2         </span></span>
<span><span class="co">#&gt;   [7] ragg_1.2.3                  tidyr_1.2.1                </span></span>
<span><span class="co">#&gt;   [9] ggplot2_3.3.6               bit64_4.0.5                </span></span>
<span><span class="co">#&gt;  [11] knitr_1.40                  R.utils_2.12.0             </span></span>
<span><span class="co">#&gt;  [13] irlba_2.3.5.1               DelayedArray_0.22.0        </span></span>
<span><span class="co">#&gt;  [15] data.table_1.14.2           rpart_4.1.16               </span></span>
<span><span class="co">#&gt;  [17] KEGGREST_1.36.3             RCurl_1.98-1.9             </span></span>
<span><span class="co">#&gt;  [19] AnnotationFilter_1.20.0     generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [21] BiocGenerics_0.42.0         GenomicFeatures_1.48.4     </span></span>
<span><span class="co">#&gt;  [23] ScaledMatrix_1.4.1          cowplot_1.1.1              </span></span>
<span><span class="co">#&gt;  [25] RSQLite_2.2.18              EnsDb.Mmusculus.v79_2.99.0 </span></span>
<span><span class="co">#&gt;  [27] RANN_2.6.1                  future_1.28.0              </span></span>
<span><span class="co">#&gt;  [29] bit_4.0.4                   spatstat.data_3.0-0        </span></span>
<span><span class="co">#&gt;  [31] xml2_1.3.3                  httpuv_1.6.6               </span></span>
<span><span class="co">#&gt;  [33] SummarizedExperiment_1.26.1 assertthat_0.2.1           </span></span>
<span><span class="co">#&gt;  [35] xfun_0.33                   hms_1.1.2                  </span></span>
<span><span class="co">#&gt;  [37] jquerylib_0.1.4             babelgene_22.9             </span></span>
<span><span class="co">#&gt;  [39] evaluate_0.17               promises_1.2.0.1           </span></span>
<span><span class="co">#&gt;  [41] fansi_1.0.3                 restfulr_0.0.15            </span></span>
<span><span class="co">#&gt;  [43] progress_1.2.2              caTools_1.18.2             </span></span>
<span><span class="co">#&gt;  [45] dbplyr_2.2.1                igraph_1.3.5               </span></span>
<span><span class="co">#&gt;  [47] DBI_1.1.3                   htmlwidgets_1.5.4          </span></span>
<span><span class="co">#&gt;  [49] spatstat.geom_3.0-3         stats4_4.2.1               </span></span>
<span><span class="co">#&gt;  [51] purrr_0.3.5                 ellipsis_0.3.2             </span></span>
<span><span class="co">#&gt;  [53] RSpectra_0.16-1             dplyr_1.0.10               </span></span>
<span><span class="co">#&gt;  [55] bookdown_0.30               annotate_1.74.0            </span></span>
<span><span class="co">#&gt;  [57] RcppParallel_5.1.5          biomaRt_2.52.0             </span></span>
<span><span class="co">#&gt;  [59] deldir_1.0-6                sparseMatrixStats_1.8.0    </span></span>
<span><span class="co">#&gt;  [61] MatrixGenerics_1.8.1        vctrs_0.4.2                </span></span>
<span><span class="co">#&gt;  [63] SingleCellExperiment_1.18.1 Biobase_2.56.0             </span></span>
<span><span class="co">#&gt;  [65] ensembldb_2.20.2            ROCR_1.0-11                </span></span>
<span><span class="co">#&gt;  [67] abind_1.4-5                 withr_2.5.0                </span></span>
<span><span class="co">#&gt;  [69] cachem_1.0.6                SIMLR_1.22.0               </span></span>
<span><span class="co">#&gt;  [71] progressr_0.11.0            sctransform_0.3.5          </span></span>
<span><span class="co">#&gt;  [73] GenomicAlignments_1.32.1    prettyunits_1.1.1          </span></span>
<span><span class="co">#&gt;  [75] scran_1.24.1                goftest_1.2-3              </span></span>
<span><span class="co">#&gt;  [77] cluster_2.1.4               ape_5.6-2                  </span></span>
<span><span class="co">#&gt;  [79] lazyeval_0.2.2              crayon_1.5.2               </span></span>
<span><span class="co">#&gt;  [81] spatstat.explore_3.0-5      edgeR_3.38.4               </span></span>
<span><span class="co">#&gt;  [83] pkgconfig_2.0.3             GenomeInfoDb_1.32.4        </span></span>
<span><span class="co">#&gt;  [85] nlme_3.1-160                ProtGenerics_1.28.0        </span></span>
<span><span class="co">#&gt;  [87] rlang_1.0.6                 globals_0.16.1             </span></span>
<span><span class="co">#&gt;  [89] lifecycle_1.0.3             miniUI_0.1.1.1             </span></span>
<span><span class="co">#&gt;  [91] filelock_1.0.2              BiocFileCache_2.4.0        </span></span>
<span><span class="co">#&gt;  [93] rsvd_1.0.5                  rprojroot_2.0.3            </span></span>
<span><span class="co">#&gt;  [95] polyclip_1.10-0             GSVA_1.44.5                </span></span>
<span><span class="co">#&gt;  [97] matrixStats_0.62.0          lmtest_0.9-40              </span></span>
<span><span class="co">#&gt;  [99] graph_1.74.0                Matrix_1.5-3               </span></span>
<span><span class="co">#&gt; [101] Rhdf5lib_1.18.2             zoo_1.8-11                 </span></span>
<span><span class="co">#&gt; [103] ggridges_0.5.4              GlobalOptions_0.1.2        </span></span>
<span><span class="co">#&gt; [105] pheatmap_1.0.12             png_0.1-7                  </span></span>
<span><span class="co">#&gt; [107] viridisLite_0.4.1           rjson_0.2.21               </span></span>
<span><span class="co">#&gt; [109] bitops_1.0-7                R.oo_1.25.0                </span></span>
<span><span class="co">#&gt; [111] rhdf5filters_1.8.0          KernSmooth_2.23-20         </span></span>
<span><span class="co">#&gt; [113] SingleCellSignalR_1.8.0     Biostrings_2.64.1          </span></span>
<span><span class="co">#&gt; [115] blob_1.2.3                  DelayedMatrixStats_1.18.1  </span></span>
<span><span class="co">#&gt; [117] shape_1.4.6                 stringr_1.4.1              </span></span>
<span><span class="co">#&gt; [119] parallelly_1.32.1           spatstat.random_3.0-1      </span></span>
<span><span class="co">#&gt; [121] S4Vectors_0.34.0            beachmat_2.12.0            </span></span>
<span><span class="co">#&gt; [123] scales_1.2.1                GSEABase_1.58.0            </span></span>
<span><span class="co">#&gt; [125] memoise_2.0.1               magrittr_2.0.3             </span></span>
<span><span class="co">#&gt; [127] plyr_1.8.7                  ica_1.0-3                  </span></span>
<span><span class="co">#&gt; [129] gplots_3.1.3                zlibbioc_1.42.0            </span></span>
<span><span class="co">#&gt; [131] compiler_4.2.1              dqrng_0.3.0                </span></span>
<span><span class="co">#&gt; [133] BiocIO_1.6.0                RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt; [135] fitdistrplus_1.1-8          Rsamtools_2.12.0           </span></span>
<span><span class="co">#&gt; [137] cli_3.4.1                   XVector_0.36.0             </span></span>
<span><span class="co">#&gt; [139] listenv_0.8.0               EnsDb.Hsapiens.v79_2.99.0  </span></span>
<span><span class="co">#&gt; [141] patchwork_1.1.2             pbapply_1.5-0              </span></span>
<span><span class="co">#&gt; [143] MASS_7.3-58.1               mgcv_1.8-40                </span></span>
<span><span class="co">#&gt; [145] tidyselect_1.2.0            stringi_1.7.8              </span></span>
<span><span class="co">#&gt; [147] textshaping_0.3.6           yaml_2.3.5                 </span></span>
<span><span class="co">#&gt; [149] BiocSingular_1.12.0         locfit_1.5-9.6             </span></span>
<span><span class="co">#&gt; [151] ggrepel_0.9.1               grid_4.2.1                 </span></span>
<span><span class="co">#&gt; [153] sass_0.4.2                  tools_4.2.1                </span></span>
<span><span class="co">#&gt; [155] future.apply_1.9.1          parallel_4.2.1             </span></span>
<span><span class="co">#&gt; [157] circlize_0.4.15             rstudioapi_0.14            </span></span>
<span><span class="co">#&gt; [159] foreach_1.5.2               bluster_1.6.0              </span></span>
<span><span class="co">#&gt; [161] AUCell_1.18.1               metapod_1.4.0              </span></span>
<span><span class="co">#&gt; [163] gridExtra_2.3               Rtsne_0.16                 </span></span>
<span><span class="co">#&gt; [165] DropletUtils_1.16.0         proxyC_0.3.3               </span></span>
<span><span class="co">#&gt; [167] digest_0.6.29               BiocManager_1.30.18        </span></span>
<span><span class="co">#&gt; [169] rgeos_0.5-9                 shiny_1.7.2                </span></span>
<span><span class="co">#&gt; [171] pracma_2.4.2                Rcpp_1.0.9                 </span></span>
<span><span class="co">#&gt; [173] GenomicRanges_1.48.0        scuttle_1.6.3              </span></span>
<span><span class="co">#&gt; [175] later_1.3.0                 RcppAnnoy_0.0.19           </span></span>
<span><span class="co">#&gt; [177] httr_1.4.4                  AnnotationDbi_1.58.0       </span></span>
<span><span class="co">#&gt; [179] colorspace_2.0-3            XML_3.99-0.11              </span></span>
<span><span class="co">#&gt; [181] fs_1.5.2                    tensor_1.5                 </span></span>
<span><span class="co">#&gt; [183] reticulate_1.26             IRanges_2.30.1             </span></span>
<span><span class="co">#&gt; [185] splines_4.2.1               uwot_0.1.14                </span></span>
<span><span class="co">#&gt; [187] statmod_1.4.37              spatstat.utils_3.0-1       </span></span>
<span><span class="co">#&gt; [189] pkgdown_2.0.6               sp_1.5-0                   </span></span>
<span><span class="co">#&gt; [191] multtest_2.52.0             plotly_4.10.0              </span></span>
<span><span class="co">#&gt; [193] systemfonts_1.0.4           xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [195] jsonlite_1.8.2              R6_2.5.1                   </span></span>
<span><span class="co">#&gt; [197] pillar_1.8.1                htmltools_0.5.3            </span></span>
<span><span class="co">#&gt; [199] mime_0.12                   glue_1.6.2                 </span></span>
<span><span class="co">#&gt; [201] fastmap_1.1.0               BiocParallel_1.30.4        </span></span>
<span><span class="co">#&gt; [203] BiocNeighbors_1.14.0        codetools_0.2-18           </span></span>
<span><span class="co">#&gt; [205] utf8_1.2.2                  lattice_0.20-45            </span></span>
<span><span class="co">#&gt; [207] bslib_0.4.0                 spatstat.sparse_3.0-0      </span></span>
<span><span class="co">#&gt; [209] tibble_3.1.8                curl_4.3.3                 </span></span>
<span><span class="co">#&gt; [211] leiden_0.4.3                gtools_3.9.3               </span></span>
<span><span class="co">#&gt; [213] magick_2.7.3                survival_3.4-0             </span></span>
<span><span class="co">#&gt; [215] limma_3.52.4                rmarkdown_2.17             </span></span>
<span><span class="co">#&gt; [217] desc_1.4.2                  munsell_0.5.0              </span></span>
<span><span class="co">#&gt; [219] rhdf5_2.40.0                GenomeInfoDbData_1.2.8     </span></span>
<span><span class="co">#&gt; [221] iterators_1.0.14            HDF5Array_1.24.2           </span></span>
<span><span class="co">#&gt; [223] msigdbr_7.5.1               reshape2_1.4.4             </span></span>
<span><span class="co">#&gt; [225] gtable_0.3.1                spatstat.core_2.4-4        </span></span>
<span><span class="co">#&gt; [227] Seurat_4.2.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yue Cao, Yingxin Lin, Ellis Patrick, Pengyi Yang, Jean Yee Hwa Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
