<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scFeatures">
<title>A detailed explanation of scFeatures' features • scFeatures</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="A detailed explanation of scFeatures' features">
<meta property="og:description" content="scFeatures">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scFeatures</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/scFeatures_overview.html">A detailed explanation of scFeatures' features</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/SydneyBioX/scFeatures/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>A detailed explanation of scFeatures' features</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/SydneyBioX/scFeatures/blob/HEAD/vignettes/scFeatures_overview.Rmd" class="external-link"><code>vignettes/scFeatures_overview.Rmd</code></a></small>
      <div class="d-none name"><code>scFeatures_overview.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>scFeatures is a tool for generating multi-view representations of
samples in a single-cell dataset. This vignette provides an overview of
scFeatures. It uses the main function to generate features and then
illustrates case studies of using the generated features for
classification, survival analysis and association study.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/scFeatures/">scFeatures</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="running-scfeatures">Running scFeatures<a class="anchor" aria-label="anchor" href="#running-scfeatures"></a>
</h2>
<p>scFeatures can be run using one line of code
<code>scfeatures_result &lt;- scFeatures(data)</code>, which generates a
list of dataframes containing all feature types in the form of samples x
features.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_scrnaseq"</span> , package <span class="op">=</span> <span class="st">"scFeatures"</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">example_scrnaseq</span></span>
<span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scFeatures.html">scFeatures</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">data</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">data</span>, sample <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">sample</span>, celltype <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">celltype</span>,</span>
<span>                                feature_types <span class="op">=</span> <span class="st">"gene_mean_celltype"</span>  , </span>
<span>                                type <span class="op">=</span> <span class="st">"scrna"</span>,  </span>
<span>                                ncores <span class="op">=</span> <span class="fl">1</span>,  </span>
<span>                                species <span class="op">=</span> <span class="st">"Homo sapiens"</span><span class="op">)</span></span></code></pre></div>
<p>By default, the above function generates all feature types. To reduce
the computational time for the demonstrate, here we generate only the
selected feature type “gene mean celltype”. More information on the
function customisation can be obtained by typing
<code>?scFeatures()</code></p>
<div class="section level3">
<h3 id="classification-of-conditions-using-the-generated-features">Classification of conditions using the generated features<a class="anchor" aria-label="anchor" href="#classification-of-conditions-using-the-generated-features"></a>
</h3>
<p>To build disease prediction model from the generated features we
utilise <a href="https://www.bioconductor.org/packages/release/bioc/html/ClassifyR.html" class="external-link"><code>ClassifyR</code></a>.</p>
<p>The output from scFeatures is a matrix of sample x feature, ie, the
row corresponds to each sample, the column corresponds to the feature,
and can be directly used as the <code>X</code>. The order of the rows is
in the order of unique(data$sample).</p>
<p>Here we use the feature type gene mean celltype as an example to
build classification model on the disease condition.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> </span>
<span><span class="va">feature_gene_mean_celltype</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result</span><span class="op">$</span><span class="va">gene_mean_celltype</span></span>
<span></span>
<span><span class="co"># inspect the first 5 rows and first 5 columns</span></span>
<span><span class="va">feature_gene_mean_celltype</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;         Naive T Cells--SNRPD2 Naive T Cells--NOSIP Naive T Cells--IL2RG</span></span>
<span><span class="co">#&gt; Pre_P8              1.6774074             1.856023             1.834704</span></span>
<span><span class="co">#&gt; Pre_P6              3.4815573             3.217231             3.583760</span></span>
<span><span class="co">#&gt; Pre_P27             0.0000000             1.486346             1.742069</span></span>
<span><span class="co">#&gt; Pre_P7              1.1761242             1.282083             2.050024</span></span>
<span><span class="co">#&gt; Pre_P20             0.6999054             1.315393             2.781787</span></span>
<span><span class="co">#&gt;         Naive T Cells--ALDOA Naive T Cells--LUC7L3</span></span>
<span><span class="co">#&gt; Pre_P8              1.598921             1.8056758</span></span>
<span><span class="co">#&gt; Pre_P6              3.493135             0.0000000</span></span>
<span><span class="co">#&gt; Pre_P27             3.518687             1.5080699</span></span>
<span><span class="co">#&gt; Pre_P7              2.575957             0.9163035</span></span>
<span><span class="co">#&gt; Pre_P20             1.626403             1.2659969</span></span>
<span></span>
<span><span class="co"># inspect the dimension of the matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">feature_gene_mean_celltype</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]   19 2217</span></span></code></pre></div>
<p>We recommend using <code><a href="https://sydneybiox.github.io/ClassifyR/reference/crossValidate.html" class="external-link">ClassifyR::crossValidate</a></code> to do
cross-validated classification with the extracted feaures.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/ClassifyR/" class="external-link">ClassifyR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># X is the feature type generated</span></span>
<span><span class="co"># y is the condition for classification</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="va">feature_gene_mean_celltype</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span>, <span class="va">y</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span>, <span class="op">]</span><span class="op">$</span><span class="va">condition</span></span>
<span></span>
<span><span class="co"># run the classification model using random forest</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">ClassifyR</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/crossValidate.html" class="external-link">crossValidate</a></span><span class="op">(</span></span>
<span>    <span class="va">X</span>, <span class="va">y</span>,</span>
<span>    classifier <span class="op">=</span> <span class="st">"randomForest"</span>, nCores <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    nFolds <span class="op">=</span> <span class="fl">3</span>, nRepeats <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ClassifyR</span><span class="fu">::</span><span class="fu"><a href="https://sydneybiox.github.io/ClassifyR/reference/performancePlot.html" class="external-link">performancePlot</a></span><span class="op">(</span>results <span class="op">=</span> <span class="va">result</span><span class="op">)</span></span></code></pre></div>
<p><img src="scFeatures_overview_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
<p>It is expected that the classification accuracy is low. This is
because we are using a small subset of data containing only 3523 genes
and 519 cells. The dataset is unlikely to contain enough information to
distinguish responders and non-responders.</p>
</div>
<div class="section level3">
<h3 id="survival-analysis-using-the-generated-features">Survival analysis using the generated features<a class="anchor" aria-label="anchor" href="#survival-analysis-using-the-generated-features"></a>
</h3>
<p>Suppose we want to use the features to perform survival analysis. In
here, since the patient outcomes are responder and non-responder, and do
not contain survival information, we randomly “generate” the survival
outcome for the purpose of demonstration.</p>
<p>We use a standard hierarchical clustering to split the patients into
2 groups based on the generated features.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/therneau/survival" class="external-link">survival</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/survminer/index.html" class="external-link">survminer</a></span><span class="op">)</span></span>
<span> </span>
<span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="va">feature_gene_mean_celltype</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run hierarchical clustering</span></span>
<span><span class="va">hclust_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">as.dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">X</span>, method <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"ward.D2"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># generate some survival outcome, including the survival days and the censoring outcome</span></span>
<span><span class="va">survival_day</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">censoring</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cluster_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html" class="external-link">cutree</a></span><span class="op">(</span><span class="va">hclust_res</span>, k <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cluster_res</span><span class="op">)</span>,</span>
<span>                        survival_day <span class="op">=</span> <span class="va">survival_day</span>,</span>
<span>                        censoring <span class="op">=</span> <span class="va">censoring</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot survival curve</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/survival/man/survfit.html" class="external-link">survfit</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/survival/man/Surv.html" class="external-link">Surv</a></span><span class="op">(</span><span class="va">survival_day</span>, <span class="va">censoring</span><span class="op">)</span> <span class="op">~</span> <span class="va">cluster</span>,</span>
<span>    data <span class="op">=</span> <span class="va">metadata</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ggsurv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/survminer/man/ggsurvplot.html" class="external-link">ggsurvplot</a></span><span class="op">(</span><span class="va">fit</span>,</span>
<span>    conf.int <span class="op">=</span> <span class="cn">FALSE</span>, risk.table <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    risk.table.col <span class="op">=</span> <span class="st">"strata"</span>, pval <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ggsurv</span></span></code></pre></div>
<p><img src="scFeatures_overview_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
<p>The p-value is very high, indicating there is not enough evidence to
claim there is a survival difference between the two groups. This is as
expected, because we randomly assigned survival status to each of the
patient.</p>
</div>
</div>
<div class="section level2">
<h2 id="association-study-of-the-features-with-the-conditions">Association study of the features with the conditions<a class="anchor" aria-label="anchor" href="#association-study-of-the-features-with-the-conditions"></a>
</h2>
<p>scFeatures provides a function that automatically run association
study of the features with the conditions and produce an HTML file with
the visualisation of the features and the association result.</p>
<p>For this, we would first need to generate the features using
scFeatures and then store the result in a named list format.</p>
<p>For demonstration purpose, we provide an example of this features
list. The code below show the steps of generating the HTML output from
the features list.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># here we use the demo data from the package </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"scfeatures_result"</span> , package <span class="op">=</span> <span class="st">"scFeatures"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># here we use the current working directory to save the html output</span></span>
<span><span class="co"># modify this to save the html file to other directory</span></span>
<span><span class="va">output_folder</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/run_association_study_report.html">run_association_study_report</a></span><span class="op">(</span><span class="va">scfeatures_result</span>, <span class="va">output_folder</span> <span class="op">)</span></span>
<span><span class="co">#&gt; /usr/lib/rstudio-server/bin/quarto/bin/tools/pandoc +RTS -K512m -RTS output_report.knit.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output /tmp/RtmpKrSVvj/output_report.html --lua-filter /enna/nobackup/yuec/R/rmarkdown/rmarkdown/lua/pagebreak.lua --lua-filter /enna/nobackup/yuec/R/rmarkdown/rmarkdown/lua/latex-div.lua --embed-resources --standalone --variable bs3=TRUE --section-divs --table-of-contents --toc-depth 3 --variable toc_float=1 --variable toc_selectors=h1,h2,h3 --variable toc_collapsed=1 --variable toc_print=1 --template /enna/nobackup/yuec/R/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --number-sections --variable theme=bootstrap --mathjax --variable 'mathjax-url=https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --include-in-header /tmp/RtmpKrSVvj/rmarkdown-str1260a933756315.html --variable code_folding=hide --variable code_menu=1</span></span></code></pre></div>
<p>Inside the directory defined in the <code>output_folder</code>, you
will see the html report output with the name
<code>output_report.html</code>.</p>
</div>
<div class="section level2">
<h2 id="sessioninfo">sessionInfo()<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Debian GNU/Linux 12 (bookworm)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.21.so;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Australia/Sydney</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] grid     graphics stats4   stats    utils    methods  base    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] data.table_1.14.8           enrichplot_1.20.0          </span></span>
<span><span class="co">#&gt;  [3] DOSE_3.26.1                 clusterProfiler_4.8.1      </span></span>
<span><span class="co">#&gt;  [5] msigdbr_7.5.1               EnsDb.Hsapiens.v79_2.99.0  </span></span>
<span><span class="co">#&gt;  [7] ensembldb_2.24.0            AnnotationFilter_1.24.0    </span></span>
<span><span class="co">#&gt;  [9] GenomicFeatures_1.52.1      org.Hs.eg.db_3.17.0        </span></span>
<span><span class="co">#&gt; [11] AnnotationDbi_1.62.2        plotly_4.10.3              </span></span>
<span><span class="co">#&gt; [13] igraph_1.5.1                tidyr_1.3.0                </span></span>
<span><span class="co">#&gt; [15] DT_0.30                     limma_3.56.2               </span></span>
<span><span class="co">#&gt; [17] pheatmap_1.0.12             dplyr_1.1.3                </span></span>
<span><span class="co">#&gt; [19] reshape2_1.4.4              survminer_0.4.9            </span></span>
<span><span class="co">#&gt; [21] ggpubr_0.6.0                ggplot2_3.4.4              </span></span>
<span><span class="co">#&gt; [23] ClassifyR_3.7.2             survival_3.5-7             </span></span>
<span><span class="co">#&gt; [25] BiocParallel_1.34.2         MultiAssayExperiment_1.26.0</span></span>
<span><span class="co">#&gt; [27] SummarizedExperiment_1.30.2 Biobase_2.60.0             </span></span>
<span><span class="co">#&gt; [29] GenomicRanges_1.52.0        GenomeInfoDb_1.36.1        </span></span>
<span><span class="co">#&gt; [31] IRanges_2.34.1              MatrixGenerics_1.12.2      </span></span>
<span><span class="co">#&gt; [33] matrixStats_1.0.0           generics_0.1.3             </span></span>
<span><span class="co">#&gt; [35] SeuratObject_5.0.0          sp_2.1-1                   </span></span>
<span><span class="co">#&gt; [37] scFeatures_1.3.3            S4Vectors_0.38.1           </span></span>
<span><span class="co">#&gt; [39] BiocGenerics_0.46.0         BiocStyle_2.28.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] R.methodsS3_1.8.2           GSEABase_1.62.0            </span></span>
<span><span class="co">#&gt;   [3] progress_1.2.2              EnsDb.Mmusculus.v79_2.99.0 </span></span>
<span><span class="co">#&gt;   [5] goftest_1.2-3               Biostrings_2.68.1          </span></span>
<span><span class="co">#&gt;   [7] HDF5Array_1.28.1            vctrs_0.6.4                </span></span>
<span><span class="co">#&gt;   [9] spatstat.random_3.2-1       digest_0.6.33              </span></span>
<span><span class="co">#&gt;  [11] png_0.1-8                   shape_1.4.6                </span></span>
<span><span class="co">#&gt;  [13] ggrepel_0.9.4               deldir_1.0-9               </span></span>
<span><span class="co">#&gt;  [15] parallelly_1.36.0           MASS_7.3-60                </span></span>
<span><span class="co">#&gt;  [17] pkgdown_2.0.7               httpuv_1.6.12              </span></span>
<span><span class="co">#&gt;  [19] foreach_1.5.2               qvalue_2.32.0              </span></span>
<span><span class="co">#&gt;  [21] withr_2.5.1                 ggfun_0.1.3                </span></span>
<span><span class="co">#&gt;  [23] xfun_0.40                   ellipsis_0.3.2             </span></span>
<span><span class="co">#&gt;  [25] memoise_2.0.1               proxyC_0.3.4               </span></span>
<span><span class="co">#&gt;  [27] commonmark_1.9.0            gson_0.1.0                 </span></span>
<span><span class="co">#&gt;  [29] systemfonts_1.0.5           tidytree_0.4.5             </span></span>
<span><span class="co">#&gt;  [31] ragg_1.2.6                  zoo_1.8-12                 </span></span>
<span><span class="co">#&gt;  [33] GlobalOptions_0.1.2         gtools_3.9.4               </span></span>
<span><span class="co">#&gt;  [35] SingleCellSignalR_1.12.0    R.oo_1.25.0                </span></span>
<span><span class="co">#&gt;  [37] prettyunits_1.2.0           KEGGREST_1.40.0            </span></span>
<span><span class="co">#&gt;  [39] promises_1.2.1              httr_1.4.7                 </span></span>
<span><span class="co">#&gt;  [41] downloader_0.4              rstatix_0.7.2              </span></span>
<span><span class="co">#&gt;  [43] restfulr_0.0.15             globals_0.16.2             </span></span>
<span><span class="co">#&gt;  [45] rhdf5filters_1.12.1         rhdf5_2.44.0               </span></span>
<span><span class="co">#&gt;  [47] rstudioapi_0.15.0           babelgene_22.9             </span></span>
<span><span class="co">#&gt;  [49] curl_5.0.2                  zlibbioc_1.46.0            </span></span>
<span><span class="co">#&gt;  [51] ScaledMatrix_1.8.1          ggraph_2.1.0               </span></span>
<span><span class="co">#&gt;  [53] polyclip_1.10-6             GenomeInfoDbData_1.2.10    </span></span>
<span><span class="co">#&gt;  [55] xtable_1.8-4                stringr_1.5.0              </span></span>
<span><span class="co">#&gt;  [57] desc_1.4.2                  evaluate_0.22              </span></span>
<span><span class="co">#&gt;  [59] S4Arrays_1.0.6              BiocFileCache_2.8.0        </span></span>
<span><span class="co">#&gt;  [61] hms_1.1.3                   bookdown_0.36              </span></span>
<span><span class="co">#&gt;  [63] irlba_2.3.5.1               colorspace_2.1-0           </span></span>
<span><span class="co">#&gt;  [65] filelock_1.0.2              spatstat.data_3.0-3        </span></span>
<span><span class="co">#&gt;  [67] magrittr_2.0.3              ggtree_3.8.0               </span></span>
<span><span class="co">#&gt;  [69] viridis_0.6.4               later_1.3.1                </span></span>
<span><span class="co">#&gt;  [71] lattice_0.22-5              spatstat.geom_3.2-7        </span></span>
<span><span class="co">#&gt;  [73] future.apply_1.11.0         genefilter_1.82.1          </span></span>
<span><span class="co">#&gt;  [75] shadowtext_0.1.2            XML_3.99-0.14              </span></span>
<span><span class="co">#&gt;  [77] scuttle_1.10.1              cowplot_1.1.1              </span></span>
<span><span class="co">#&gt;  [79] ggupset_0.3.0               pillar_1.9.0               </span></span>
<span><span class="co">#&gt;  [81] nlme_3.1-163                iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [83] caTools_1.18.2              compiler_4.3.1             </span></span>
<span><span class="co">#&gt;  [85] beachmat_2.16.0             stringi_1.7.12             </span></span>
<span><span class="co">#&gt;  [87] tensor_1.5                  GenomicAlignments_1.36.0   </span></span>
<span><span class="co">#&gt;  [89] plyr_1.8.9                  crayon_1.5.2               </span></span>
<span><span class="co">#&gt;  [91] abind_1.4-5                 BiocIO_1.10.0              </span></span>
<span><span class="co">#&gt;  [93] gridGraphics_0.5-1          ggtext_0.1.2               </span></span>
<span><span class="co">#&gt;  [95] locfit_1.5-9.8              graphlayouts_1.0.1         </span></span>
<span><span class="co">#&gt;  [97] bit_4.0.5                   fastmatch_1.1-4            </span></span>
<span><span class="co">#&gt;  [99] codetools_0.2-19            textshaping_0.3.7          </span></span>
<span><span class="co">#&gt; [101] BiocSingular_1.16.0         crosstalk_1.2.0            </span></span>
<span><span class="co">#&gt; [103] bslib_0.5.1                 multtest_2.56.0            </span></span>
<span><span class="co">#&gt; [105] mime_0.12                   splines_4.3.1              </span></span>
<span><span class="co">#&gt; [107] markdown_1.11               circlize_0.4.15            </span></span>
<span><span class="co">#&gt; [109] Rcpp_1.0.11                 datasets_4.3.1             </span></span>
<span><span class="co">#&gt; [111] dbplyr_2.4.0                sparseMatrixStats_1.12.2   </span></span>
<span><span class="co">#&gt; [113] HDO.db_0.99.1               gridtext_0.1.5             </span></span>
<span><span class="co">#&gt; [115] knitr_1.44                  blob_1.2.4                 </span></span>
<span><span class="co">#&gt; [117] utf8_1.2.4                  fs_1.6.3                   </span></span>
<span><span class="co">#&gt; [119] listenv_0.9.0               DelayedMatrixStats_1.22.1  </span></span>
<span><span class="co">#&gt; [121] GSVA_1.48.2                 ggplotify_0.1.2            </span></span>
<span><span class="co">#&gt; [123] ggsignif_0.6.4              tibble_3.2.1               </span></span>
<span><span class="co">#&gt; [125] Matrix_1.6-4                statmod_1.5.0              </span></span>
<span><span class="co">#&gt; [127] tweenr_2.0.2                pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt; [129] tools_4.3.1                 cachem_1.0.8               </span></span>
<span><span class="co">#&gt; [131] RSQLite_2.3.1               viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt; [133] DBI_1.1.3                   fastmap_1.1.1              </span></span>
<span><span class="co">#&gt; [135] rmarkdown_2.25              scales_1.2.1               </span></span>
<span><span class="co">#&gt; [137] Rsamtools_2.16.0            broom_1.0.5                </span></span>
<span><span class="co">#&gt; [139] sass_0.4.7                  patchwork_1.1.3            </span></span>
<span><span class="co">#&gt; [141] BiocManager_1.30.22         dotCall64_1.1-0            </span></span>
<span><span class="co">#&gt; [143] graph_1.78.0                carData_3.0-5              </span></span>
<span><span class="co">#&gt; [145] farver_2.1.1                scatterpie_0.2.1           </span></span>
<span><span class="co">#&gt; [147] tidygraph_1.2.3             yaml_2.3.7                 </span></span>
<span><span class="co">#&gt; [149] rtracklayer_1.60.0          cli_3.6.1                  </span></span>
<span><span class="co">#&gt; [151] purrr_1.0.2                 lifecycle_1.0.3            </span></span>
<span><span class="co">#&gt; [153] bluster_1.10.0              backports_1.4.1            </span></span>
<span><span class="co">#&gt; [155] annotate_1.78.0             gtable_0.3.4               </span></span>
<span><span class="co">#&gt; [157] rjson_0.2.21                progressr_0.14.0           </span></span>
<span><span class="co">#&gt; [159] parallel_4.3.1              ape_5.7-1                  </span></span>
<span><span class="co">#&gt; [161] jsonlite_1.8.7              edgeR_3.42.4               </span></span>
<span><span class="co">#&gt; [163] bitops_1.0-7                bit64_4.0.5                </span></span>
<span><span class="co">#&gt; [165] Rtsne_0.16                  yulab.utils_0.1.0          </span></span>
<span><span class="co">#&gt; [167] spatstat.utils_3.0-4        BiocNeighbors_1.18.0       </span></span>
<span><span class="co">#&gt; [169] ranger_0.15.1               RcppParallel_5.1.7         </span></span>
<span><span class="co">#&gt; [171] jquerylib_0.1.4             metapod_1.8.0              </span></span>
<span><span class="co">#&gt; [173] GOSemSim_2.26.0             dqrng_0.3.1                </span></span>
<span><span class="co">#&gt; [175] survMisc_0.5.6              R.utils_2.12.2             </span></span>
<span><span class="co">#&gt; [177] lazyeval_0.2.2              shiny_1.7.5.1              </span></span>
<span><span class="co">#&gt; [179] htmltools_0.5.6.1           KMsurv_0.1-5               </span></span>
<span><span class="co">#&gt; [181] GO.db_3.17.0                rappdirs_0.3.3             </span></span>
<span><span class="co">#&gt; [183] glue_1.6.2                  spam_2.10-0                </span></span>
<span><span class="co">#&gt; [185] XVector_0.40.0              RCurl_1.98-1.12            </span></span>
<span><span class="co">#&gt; [187] treeio_1.24.1               rprojroot_2.0.3            </span></span>
<span><span class="co">#&gt; [189] scran_1.28.1                gridExtra_2.3              </span></span>
<span><span class="co">#&gt; [191] AUCell_1.22.0               R6_2.5.1                   </span></span>
<span><span class="co">#&gt; [193] SingleCellExperiment_1.22.0 gplots_3.1.3               </span></span>
<span><span class="co">#&gt; [195] km.ci_0.5-6                 labeling_0.4.3             </span></span>
<span><span class="co">#&gt; [197] cluster_2.1.4               Rhdf5lib_1.22.0            </span></span>
<span><span class="co">#&gt; [199] aplot_0.2.2                 DelayedArray_0.26.6        </span></span>
<span><span class="co">#&gt; [201] tidyselect_1.2.0            ProtGenerics_1.32.0        </span></span>
<span><span class="co">#&gt; [203] ggforce_0.4.1               xml2_1.3.5                 </span></span>
<span><span class="co">#&gt; [205] car_3.1-2                   future_1.33.0              </span></span>
<span><span class="co">#&gt; [207] rsvd_1.0.5                  munsell_0.5.0              </span></span>
<span><span class="co">#&gt; [209] KernSmooth_2.23-22          grDevices_4.3.1            </span></span>
<span><span class="co">#&gt; [211] htmlwidgets_1.6.2           fgsea_1.26.0               </span></span>
<span><span class="co">#&gt; [213] RColorBrewer_1.1-3          biomaRt_2.56.1             </span></span>
<span><span class="co">#&gt; [215] rlang_1.1.1                 spatstat.sparse_3.0-3      </span></span>
<span><span class="co">#&gt; [217] spatstat.explore_3.2-5      fansi_1.0.5</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yue Cao, Yingxin Lin, Ellis Patrick, Pengyi Yang, Jean Yee Hwa Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
