<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ scFeatures</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scFeatures</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/scFeatures_associationstudy.html">An association study of features with conditions with scFeatures</a>
    </li>
    <li>
      <a href="../articles/scFeatures_detail.html">A detailed explanation of scFeatures' features</a>
    </li>
    <li>
      <a href="../articles/scFeatures_summary.html">An overview of scFeatures' functions</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/SydneyBioX/scFeatures/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="L_stats.html">L_stats()</a></code> </p>
        </td>
        <td><p>Compute L statistic for a point pattern</p></td>
      </tr><tr><td>
          <p><code><a href="bulk_sample.html">bulk_sample()</a></code> </p>
        </td>
        <td><p>Create pseudo-bulk for each sample in a Seurat object</p></td>
      </tr><tr><td>
          <p><code><a href="bulk_sample_celltype.html">bulk_sample_celltype()</a></code> </p>
        </td>
        <td><p>Create pseudo-bulk for each cell type of each sample in a Seurat object</p></td>
      </tr><tr><td>
          <p><code><a href="check_data.html">check_data()</a></code> </p>
        </td>
        <td><p>Check if required metadata is in the Seurat object</p></td>
      </tr><tr><td>
          <p><code><a href="generateBPParam.html">generateBPParam()</a></code> </p>
        </td>
        <td><p>Enable parallel processing</p></td>
      </tr><tr><td>
          <p><code><a href="get_num_cell_per_celltype.html">get_num_cell_per_celltype()</a></code> </p>
        </td>
        <td><p>Compute number of cells in each cell type in each spot
for spatial transcriptomics data</p></td>
      </tr><tr><td>
          <p><code><a href="get_num_cell_per_spot.html">get_num_cell_per_spot()</a></code> </p>
        </td>
        <td><p>Estimate a relative number of cells per spot for
spatial transcriptomics data</p></td>
      </tr><tr><td>
          <p><code><a href="individual_geneset_proportion_celltype.html">individual_geneset_proportion_celltype()</a></code> </p>
        </td>
        <td><p>helper function to calculate proportion that a gene is expressed in each
cell type in a sample</p></td>
      </tr><tr><td>
          <p><code><a href="makeSeurat.html">makeSeurat()</a></code> </p>
        </td>
        <td><p>Format SingleCellExperiment or SpatialExperiment object into Seurat object
for scFeatures functions</p></td>
      </tr><tr><td>
          <p><code><a href="process_data.html">process_data()</a></code> </p>
        </td>
        <td><p>data pre-processing</p></td>
      </tr><tr><td>
          <p><code><a href="rearrange_string.html">rearrange_string()</a></code> </p>
        </td>
        <td><p>Rearrange string</p></td>
      </tr><tr><td>
          <p><code><a href="remove_mito.html">remove_mito()</a></code> </p>
        </td>
        <td><p>remove mito and ribo genes and genes highly correlated with mito genes
as mitochondria and ribosomal genes may not as interesting to look at</p></td>
      </tr><tr><td>
          <p><code><a href="run_L_function.html">run_L_function()</a></code> </p>
        </td>
        <td><p>Generate L stats</p></td>
      </tr><tr><td>
          <p><code><a href="run_Morans_I.html">run_Morans_I()</a></code> </p>
        </td>
        <td><p>generate Moran's I</p></td>
      </tr><tr><td>
          <p><code><a href="run_association_study_report.html">run_association_study_report()</a></code> </p>
        </td>
        <td><p>Create an association study report in HTML format</p></td>
      </tr><tr><td>
          <p><code><a href="run_celltype_interaction.html">run_celltype_interaction()</a></code> </p>
        </td>
        <td><p>generate cell type interaction</p></td>
      </tr><tr><td>
          <p><code><a href="run_gene_cor.html">run_gene_cor()</a></code> </p>
        </td>
        <td><p>Generate overall aggregated gene correlation</p></td>
      </tr><tr><td>
          <p><code><a href="run_gene_cor_celltype.html">run_gene_cor_celltype()</a></code> </p>
        </td>
        <td><p>Generate cell type specific gene expression correlation</p></td>
      </tr><tr><td>
          <p><code><a href="run_gene_mean.html">run_gene_mean()</a></code> </p>
        </td>
        <td><p>Generate overall aggregated mean expression</p></td>
      </tr><tr><td>
          <p><code><a href="run_gene_mean_celltype.html">run_gene_mean_celltype()</a></code> </p>
        </td>
        <td><p>Generate cell type specific gene mean expression</p></td>
      </tr><tr><td>
          <p><code><a href="run_gene_prop.html">run_gene_prop()</a></code> </p>
        </td>
        <td><p>Generate overall aggregated gene proportion expression</p></td>
      </tr><tr><td>
          <p><code><a href="run_gene_prop_celltype.html">run_gene_prop_celltype()</a></code> </p>
        </td>
        <td><p>Generate cell type specific gene proportion expression</p></td>
      </tr><tr><td>
          <p><code><a href="run_nn_correlation.html">run_nn_correlation()</a></code> </p>
        </td>
        <td><p>generate nearest neighbour correlation</p></td>
      </tr><tr><td>
          <p><code><a href="run_pathway_gsva.html">run_pathway_gsva()</a></code> </p>
        </td>
        <td><p>Generate pathway score using gene set enrichement analysis</p></td>
      </tr><tr><td>
          <p><code><a href="run_pathway_mean.html">run_pathway_mean()</a></code> </p>
        </td>
        <td><p>Generate pathway score using expression level</p></td>
      </tr><tr><td>
          <p><code><a href="run_pathway_prop.html">run_pathway_prop()</a></code> </p>
        </td>
        <td><p>Generate pathway score using proportion of expression</p></td>
      </tr><tr><td>
          <p><code><a href="run_proportion_logit.html">run_proportion_logit()</a></code> </p>
        </td>
        <td><p>Generate cell type proportions, with logit transformation</p></td>
      </tr><tr><td>
          <p><code><a href="run_proportion_ratio.html">run_proportion_ratio()</a></code> </p>
        </td>
        <td><p>Generate cell type proportion ratio</p></td>
      </tr><tr><td>
          <p><code><a href="run_proportion_raw.html">run_proportion_raw()</a></code> </p>
        </td>
        <td><p>Generate cell type proportion raw</p></td>
      </tr><tr><td>
          <p><code><a href="scFeatures.html">scFeatures()</a></code> </p>
        </td>
        <td><p>wrapper function to run all feature types in scFeatures</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yue Cao, Yingxin Lin, Ellis Patrick, Pengyi Yang, Jean Yee Hwa Yang.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

